<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.340">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Tutorial: Generating substrate specific thiolases with PoET </title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./favicon.svg" rel="icon" type="image/svg+xml">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./prots2prot-introduction-page.html">Prots2prot</a></li><li class="breadcrumb-item"><a href="./prots2prot-thiolase-codehidden.html">Tutorial: Generating substrate specific thiolases with PoET <a class="anchor" id="first"></a></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./logo.svg" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Basics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to OpenProtein</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./account-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Account</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Project</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dataset-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Dataset</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./visualizations-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Visualizations</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./job-status-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Job Status</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./predictions-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Predictions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./design-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Design</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./library-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Libraries</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Models</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./models-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Training model(s)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cross-validation-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Model diagnostics and cross validation</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Walkthroughs</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./demo-datasets-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Demo datasets</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mutagenesis-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ML-guided mutagenesis and sequence-to-function modeling walkthrough</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Prots2prot</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prots2prot-introduction-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to Prots2prot</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prots2prot-defining-prompts-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Using Prots2prot</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prots2prot-tools-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prots2prot Tools</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prots2prot-tutorial-page.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tutorial: Designing de novo variant libraries using Prots2prot</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prots2prot-thiolase-codehidden.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Tutorial: Generating substrate specific thiolases with PoET <a class="anchor" id="first"></a></span></a><a class="anchor" id="first"></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#thiolase-superfamily" id="toc-thiolase-superfamily" class="nav-link active" data-scroll-target="#thiolase-superfamily">Thiolase superfamily </a><a class="anchor nav-link" id="1_0" data-scroll-target="undefined"></a>
  <ul class="collapse">
  <li><a href="#activity-and-specificity-dataset" id="toc-activity-and-specificity-dataset" class="nav-link" data-scroll-target="#activity-and-specificity-dataset">Activity and specificity dataset </a><a class="anchor nav-link" id="1_1" data-scroll-target="undefined"></a></li>
  <li><a href="#loading-the-data" id="toc-loading-the-data" class="nav-link" data-scroll-target="#loading-the-data">Loading the data</a><a class="anchor nav-link" id="1_2" data-scroll-target="undefined"></a></li>
  <li><a href="#exploring-the-data" id="toc-exploring-the-data" class="nav-link" data-scroll-target="#exploring-the-data">Exploring the data </a><a class="anchor nav-link" id="1_3" data-scroll-target="undefined"></a></li>
  <li><a href="#choosing-a-target-candidate" id="toc-choosing-a-target-candidate" class="nav-link" data-scroll-target="#choosing-a-target-candidate">Choosing a target candidate</a><a class="anchor nav-link" id="1_4" data-scroll-target="undefined"></a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary">Summary </a><a class="anchor nav-link" id="1_5" data-scroll-target="undefined"></a></li>
  </ul></li>
  <li><a href="#poet-thiolase-scoring" id="toc-poet-thiolase-scoring" class="nav-link" data-scroll-target="#poet-thiolase-scoring">PoET thiolase scoring </a><a class="anchor nav-link" id="2_0" data-scroll-target="undefined"></a>
  <ul class="collapse">
  <li><a href="#job-creation" id="toc-job-creation" class="nav-link" data-scroll-target="#job-creation">Job creation </a><a class="anchor nav-link" id="2_1" data-scroll-target="undefined"></a></li>
  <li><a href="#prompting-poet" id="toc-prompting-poet" class="nav-link" data-scroll-target="#prompting-poet">Prompting PoET </a><a class="anchor nav-link" id="2_2" data-scroll-target="undefined"></a></li>
  <li><a href="#scoring-results" id="toc-scoring-results" class="nav-link" data-scroll-target="#scoring-results">Scoring results </a><a class="anchor nav-link" id="2_3" data-scroll-target="undefined"></a></li>
  <li><a href="#poet-scoring-summary" id="toc-poet-scoring-summary" class="nav-link" data-scroll-target="#poet-scoring-summary">PoET scoring Summary </a><a class="anchor nav-link" id="2_4" data-scroll-target="undefined"></a></li>
  </ul></li>
  <li><a href="#poet-de-novo-thiolase-generation" id="toc-poet-de-novo-thiolase-generation" class="nav-link" data-scroll-target="#poet-de-novo-thiolase-generation">PoET de novo thiolase generation </a><a class="anchor nav-link" id="3_0" data-scroll-target="undefined"></a>
  <ul class="collapse">
  <li><a href="#job-creation-1" id="toc-job-creation-1" class="nav-link" data-scroll-target="#job-creation-1">Job creation </a><a class="anchor nav-link" id="3_1" data-scroll-target="undefined"></a></li>
  <li><a href="#evaluating-generated-sequences" id="toc-evaluating-generated-sequences" class="nav-link" data-scroll-target="#evaluating-generated-sequences">Evaluating generated sequences </a><a class="anchor nav-link" id="3_2" data-scroll-target="undefined"></a></li>
  <li><a href="#investigating-generated-thiolase-structures" id="toc-investigating-generated-thiolase-structures" class="nav-link" data-scroll-target="#investigating-generated-thiolase-structures">Investigating generated thiolase structures</a></li>
  <li><a href="#structural-analysis" id="toc-structural-analysis" class="nav-link" data-scroll-target="#structural-analysis">Structural analysis</a></li>
  <li><a href="#generated-sequences-summary" id="toc-generated-sequences-summary" class="nav-link" data-scroll-target="#generated-sequences-summary">Generated sequences summary </a><a class="anchor nav-link" id="3_3" data-scroll-target="undefined"></a></li>
  </ul></li>
  <li><a href="#summary-1" id="toc-summary-1" class="nav-link" data-scroll-target="#summary-1">Summary </a><a class="anchor nav-link" id="4_1" data-scroll-target="undefined"></a></li>
  <li><a href="#appendix" id="toc-appendix" class="nav-link" data-scroll-target="#appendix">Appendix </a><a class="anchor nav-link" id="Appendix" data-scroll-target="undefined"></a>
  <ul class="collapse">
  <li><a href="#additional-figures" id="toc-additional-figures" class="nav-link" data-scroll-target="#additional-figures">Additional figures:</a>
  <ul class="collapse">
  <li><a href="#full-results-for-compound-activity-vs-model-score" id="toc-full-results-for-compound-activity-vs-model-score" class="nav-link" data-scroll-target="#full-results-for-compound-activity-vs-model-score">Full results for compound activity vs model score:</a></li>
  </ul></li>
  <li><a href="#additional-prompts" id="toc-additional-prompts" class="nav-link" data-scroll-target="#additional-prompts">Additional prompts:</a>
  <ul class="collapse">
  <li><a href="#scoring-with-default-parameters-yields-a-weaker-model" id="toc-scoring-with-default-parameters-yields-a-weaker-model" class="nav-link" data-scroll-target="#scoring-with-default-parameters-yields-a-weaker-model">Scoring with default parameters yields a weaker model: </a><a class="anchor nav-link" id="defaultprompt" data-scroll-target="undefined"></a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Tutorial: Generating substrate specific thiolases with PoET <a class="anchor" id="first"></a></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>This tutorial will introduce OpenProtein’s new PoET workflow, a cutting-edge protein language model designed to facilitate evaluating protein sequences from scratch, eliminating the need for functional measurements of your specific protein. With this model, you can effortlessly score candidate proteins, or generate a custom diverse library before running any wetlab experiments!</p>
<p>The core workflow in PoET is governed by a prompt, which consists of a collection of sequences that encode valuable insights into the local fitness landscape and co-evolutionary patterns associated with your protein of interest. One effective approach for constructing this prompt involves, first; generating a multiple sequence alignment (MSA) of your protein with sequences from various evolutionary lineages, and second; applying filters to that MSA to create a focussed prompt. Since these sequences share an evolutionary connection, they inherently carry information about the protein’s local fitness landscape. PoET leverages this information to both: * 1) assess the probability of observing a specific sequence based on the inferred evolutionary process * 2) generate novel sequences based on the same inferred evolutionary landscape.</p>
<p>We will apply this new machine learning model to develop novel thiolase enzymes with activity against a compound of interest.</p>
<section id="thiolase-superfamily" class="level2">
<h2 class="anchored" data-anchor-id="thiolase-superfamily">Thiolase superfamily <a class="anchor" id="1_0"></a></h2>
<p>Thiolases play a critical role in catalyzing thioester formation — a key process in fatty acid biosynthesis <em>in vivo</em>. Thiolases have attracted significant research interest due to their role in catalysing thioester formation, which is a key intermediary in several industrial processes, including: * Pharmaceuticals: thioesters can be extensively modified with specific moieties, enabling the synthesis of a wide range of pharmaceuticals with desired properties * Biopolymers: thiolases enable the efficient conversion of organic acids into thioesters, which are subsequently polymerized to create environmentally friendly biopolymers. For example polyhydroxyalkanoate (PHA). * Flavours and fragrences: thioesters are key components in the synthesis of aroma compounds and flavour enhancers</p>
<p>Understanding the factors that govern thiolase substrate specificity is therefore crucial. Furthermore, developing novel thiolases with improved specificity or activity may yield novel products or more efficient industrial processes.</p>
<p>This tutorial will build on the work and data made available by <a href="https://academic.oup.com/synbio/article/5/1/ysaa004/5847618">Robinson (2020)</a>. Robinson et al.&nbsp;investigated thiolase substrate specificity and activity with an extensive library of compounds and thiolases.</p>
<p>We will demonstrate how OpenProtein’s PoET model can leverage co-evolutionary data from an MSA to score sequences, generate new candidate sequences <em>de novo</em>, and facilitate further <em>in-vitro</em> testing. Our workflow will begin with a single seed sequence obtained freely from Uniprot, while the Robinson dataset will tie our observations back to experimental data, and provide an empirical validation of the trends we see in our modelling.</p>
<section id="activity-and-specificity-dataset" class="level3">
<h3 class="anchored" data-anchor-id="activity-and-specificity-dataset">Activity and specificity dataset <a class="anchor" id="1_1"></a></h3>
<p>I have reproduced <a href="https://academic.oup.com/view-large/figure/206116503/ysaa004f2.tif">figure 2</a> from the Robinson paper as a visual summary of the dataset; we have 15 chemical compounds of various chemical structures, and 73 thiolases from a broad range of taxonomic classes. We can note that most thiolases show activity against only a subset of compounds, and activity is highly variable (~1-3 orders of magnitude difference).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/4b207193-1-image.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">image.png</figcaption>
</figure>
</div>
<div class="cell" data-execution_count="1">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> json</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> AWSTools.Batchtools <span class="im">import</span> batch_utils <span class="im">as</span> bu </span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> openprotein</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.stats <span class="im">import</span> spearmanr</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pandas <span class="im">import</span> option_context</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> misc <span class="im">import</span>  wait_for_job</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> requests </span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>sns.set_style(<span class="st">"whitegrid"</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">'secrets.config'</span>, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>    config <span class="op">=</span> json.load(f)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>session <span class="op">=</span> openprotein.<span class="ex">connect</span>(config[<span class="st">'username'</span>], config[<span class="st">'password'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="loading-the-data" class="level3">
<h3 class="anchored" data-anchor-id="loading-the-data">Loading the data<a class="anchor" id="1_2"></a></h3>
<p>Let’s first load and examine the Robinson 2020 data:</p>
<div class="cell" data-execution_count="2">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> pd.read_csv(<span class="st">"./data/Robinson/enzyme_substrate_activity.csv"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> data.rename(columns<span class="op">=</span>{<span class="st">"cmpnd"</span>: <span class="st">"compound"</span>, <span class="st">'log_slope'</span>:<span class="st">'log_activity'</span>})</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>data[<span class="st">'compound'</span>] <span class="op">=</span> data.compound.<span class="bu">str</span>.replace(<span class="st">" "</span>,<span class="st">"_"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># read in sequences and match to data</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>sequences <span class="op">=</span> bu.fasta2csv(<span class="st">"./data/Robinson//robinson.fa"</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>sequences[<span class="st">'org'</span>] <span class="op">=</span> sequences[<span class="st">'accession'</span>].<span class="bu">apply</span>(<span class="kw">lambda</span> x: <span class="st">" "</span>.join( x.split(<span class="st">'1_'</span>)[<span class="op">-</span><span class="dv">1</span>:])   )</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>sequences[<span class="st">'org'</span>] <span class="op">=</span> sequences.org.<span class="bu">str</span>.replace(<span class="st">"_"</span>,<span class="st">" "</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>sequences <span class="op">=</span> sequences.rename(columns<span class="op">=</span>{<span class="st">"seq"</span>: <span class="st">"sequence"</span>})</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="co"># manual matches</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>sequences.loc[sequences.accession.<span class="bu">str</span>.contains(<span class="st">"Halobacteriovorax"</span>),<span class="st">'org'</span>] <span class="op">=</span> <span class="st">"Halobacteriovorax marinus"</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>sequences.loc[sequences.accession.<span class="bu">str</span>.contains(<span class="st">"Enhygromyxa"</span>),<span class="st">'org'</span>] <span class="op">=</span> <span class="st">"Enhygromyxa salina"</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>sequences.loc[sequences.accession.<span class="bu">str</span>.contains(<span class="st">"Kocuria_varians"</span>),<span class="st">'org'</span>] <span class="op">=</span> <span class="st">"Kocuria varians"</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>sequences.loc[sequences.accession.<span class="bu">str</span>.contains(<span class="st">"Pseudoxanthomonas"</span>),<span class="st">'org'</span>] <span class="op">=</span> <span class="st">"Pseudoxanthomonas"</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>sequences.loc[sequences.accession.<span class="bu">str</span>.contains(<span class="st">"translucens"</span>),<span class="st">'org'</span>] <span class="op">=</span> <span class="st">"Xanthomonas translucens"</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>sequences.loc[sequences.accession.<span class="bu">str</span>.contains(<span class="st">"Leifsonia"</span>),<span class="st">'org'</span>] <span class="op">=</span> <span class="st">"Leifsonia sp. Leaf325"</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>sequences.loc[sequences.accession.<span class="bu">str</span>.contains(<span class="st">"Chloroflexi_bacterium"</span>),<span class="st">'org'</span>] <span class="op">=</span> <span class="st">"Chloroflexi bacterium RBG 13 51 36"</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>matched <span class="op">=</span> <span class="bu">len</span>(<span class="bu">set</span>(sequences.org.unique()) <span class="op">&amp;</span> <span class="bu">set</span>(data.org.unique()))</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="co"># merge together</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> data.merge(sequences, on<span class="op">=</span><span class="st">'org'</span>, how<span class="op">=</span><span class="st">'left'</span>)</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> data[data.org.isna()].shape[<span class="dv">0</span>]<span class="op">==</span><span class="dv">0</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> data.sort_values([<span class="st">'org'</span>,<span class="st">'compound'</span>])</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>( <span class="ss">f"Matched </span><span class="sc">{</span>matched<span class="sc">}</span><span class="ss"> out of </span><span class="sc">{</span><span class="bu">len</span>(data.org.unique())<span class="sc">}</span><span class="ss"> thiolases"</span> )</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>data.head(<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Matched 72 out of 73 thiolases</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="2">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">filename</th>
<th data-quarto-table-cell-role="th">compound</th>
<th data-quarto-table-cell-role="th">org</th>
<th data-quarto-table-cell-role="th">log_activity</th>
<th data-quarto-table-cell-role="th">accession</th>
<th data-quarto-table-cell-role="th">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">404</td>
<td>output//2019-11-22/2019-11-22_7Ph heptanoate_a...</td>
<td>7Ph_heptanoate</td>
<td>Actinomycetospora chiangmaiensis</td>
<td>1.437703</td>
<td>WP_026204600.1_Actinomycetospora_chiangmaiensis</td>
<td>MRAAQRYDDVVITGLAHVDAPHVVTSTELEDRLAGTLGRLRITPGL...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">657</td>
<td>output//2019-12-13/2019-12-13_azido_all_data_c...</td>
<td>azido</td>
<td>Actinomycetospora chiangmaiensis</td>
<td>1.722814</td>
<td>WP_026204600.1_Actinomycetospora_chiangmaiensis</td>
<td>MRAAQRYDDVVITGLAHVDAPHVVTSTELEDRLAGTLGRLRITPGL...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">295</td>
<td>output//2019-11-15/2019-11-15_butoxy_all_data_...</td>
<td>butoxy</td>
<td>Actinomycetospora chiangmaiensis</td>
<td>1.989242</td>
<td>WP_026204600.1_Actinomycetospora_chiangmaiensis</td>
<td>MRAAQRYDDVVITGLAHVDAPHVVTSTELEDRLAGTLGRLRITPGL...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>We can see above the same data that corresponds to figure 2. We will need to wrangle the data a little to summarise replicates and add back in the 0 activity rows:</p>
<div class="cell" data-execution_count="3">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize replicates</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> (</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    data.groupby([<span class="st">'org'</span>, <span class="st">'compound'</span>, <span class="st">'sequence'</span>])</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    .agg(mean_log_activity<span class="op">=</span>(<span class="st">'log_activity'</span>, <span class="st">'mean'</span>),</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>         std_log_activity<span class="op">=</span>(<span class="st">'log_activity'</span>, <span class="st">'std'</span>),</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>         count_activity<span class="op">=</span>(<span class="st">'log_activity'</span>, <span class="st">'count'</span>))</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    .reset_index()</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> data.fillna(<span class="fl">0.0</span>)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="co"># add in zero-activity rows</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>new_df <span class="op">=</span> pd.DataFrame(index<span class="op">=</span>pd.MultiIndex.from_product([data[<span class="st">'org'</span>].unique(), data[<span class="st">'compound'</span>].unique()],</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>                                                       names<span class="op">=</span>[<span class="st">'org'</span>, <span class="st">'compound'</span>]))</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>new_df[<span class="st">'mean_log_activity'</span>] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>new_df[<span class="st">'std_log_activity'</span>] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="co"># update the new dataframe with the original dataframe's values</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>new_df.update(data.set_index([<span class="st">'org'</span>, <span class="st">'compound'</span>]))</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="co"># reset the index and merge</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> new_df.reset_index()</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> data.merge(sequences, on<span class="op">=</span><span class="st">'org'</span>, how<span class="op">=</span><span class="st">'left'</span>)</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="3">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">org</th>
<th data-quarto-table-cell-role="th">compound</th>
<th data-quarto-table-cell-role="th">mean_log_activity</th>
<th data-quarto-table-cell-role="th">std_log_activity</th>
<th data-quarto-table-cell-role="th">accession</th>
<th data-quarto-table-cell-role="th">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>Actinomycetospora chiangmaiensis</td>
<td>7Ph_heptanoate</td>
<td>1.437703</td>
<td>0.000000</td>
<td>WP_026204600.1_Actinomycetospora_chiangmaiensis</td>
<td>MRAAQRYDDVVITGLAHVDAPHVVTSTELEDRLAGTLGRLRITPGL...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>Actinomycetospora chiangmaiensis</td>
<td>azido</td>
<td>1.722814</td>
<td>0.000000</td>
<td>WP_026204600.1_Actinomycetospora_chiangmaiensis</td>
<td>MRAAQRYDDVVITGLAHVDAPHVVTSTELEDRLAGTLGRLRITPGL...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>Actinomycetospora chiangmaiensis</td>
<td>butoxy</td>
<td>1.989242</td>
<td>0.000000</td>
<td>WP_026204600.1_Actinomycetospora_chiangmaiensis</td>
<td>MRAAQRYDDVVITGLAHVDAPHVVTSTELEDRLAGTLGRLRITPGL...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>Actinomycetospora chiangmaiensis</td>
<td>cyclopentyl</td>
<td>1.185503</td>
<td>0.000000</td>
<td>WP_026204600.1_Actinomycetospora_chiangmaiensis</td>
<td>MRAAQRYDDVVITGLAHVDAPHVVTSTELEDRLAGTLGRLRITPGL...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>Actinomycetospora chiangmaiensis</td>
<td>heptynoate</td>
<td>1.448957</td>
<td>0.211179</td>
<td>WP_026204600.1_Actinomycetospora_chiangmaiensis</td>
<td>MRAAQRYDDVVITGLAHVDAPHVVTSTELEDRLAGTLGRLRITPGL...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div class="cell" data-execution_count="4">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> misc <span class="im">import</span> display_side_by_side</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> option_context(<span class="st">'display.max_rows'</span>, <span class="va">None</span>, <span class="st">'display.max_columns'</span>, <span class="va">None</span>):</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    display_side_by_side( [data[[<span class="st">'compound'</span>]].value_counts().head().reset_index().rename(columns<span class="op">=</span>{<span class="dv">0</span>:<span class="st">'count'</span>}), data[[<span class="st">'org'</span>]].value_counts().head().reset_index().rename(columns<span class="op">=</span>{<span class="dv">0</span>:<span class="st">'count'</span>})] , [<span class="st">'Compound'</span>, <span class="st">"Organism"</span>] )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<style type="text/css">
</style>

<table id="T_09f8c" data-quarto-postprocess="true" class="table table-sm table-striped small">
<caption>Compound</caption>
<thead>
<tr class="header">
<th class="blank level0" data-quarto-table-cell-role="th">&nbsp;</th>
<th id="T_09f8c_level0_col0" class="col_heading level0 col0" data-quarto-table-cell-role="th">compound</th>
<th id="T_09f8c_level0_col1" class="col_heading level0 col1" data-quarto-table-cell-role="th">count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td id="T_09f8c_level0_row0" class="row_heading level0 row0" data-quarto-table-cell-role="th">0</td>
<td id="T_09f8c_row0_col0" class="data row0 col0">7Ph_heptanoate</td>
<td id="T_09f8c_row0_col1" class="data row0 col1">72</td>
</tr>
<tr class="even">
<td id="T_09f8c_level0_row1" class="row_heading level0 row1" data-quarto-table-cell-role="th">1</td>
<td id="T_09f8c_row1_col0" class="data row1 col0">ClPh_propionate</td>
<td id="T_09f8c_row1_col1" class="data row1 col1">72</td>
</tr>
<tr class="odd">
<td id="T_09f8c_level0_row2" class="row_heading level0 row2" data-quarto-table-cell-role="th">2</td>
<td id="T_09f8c_row2_col0" class="data row2 col0">TMA</td>
<td id="T_09f8c_row2_col1" class="data row2 col1">72</td>
</tr>
<tr class="even">
<td id="T_09f8c_level0_row3" class="row_heading level0 row3" data-quarto-table-cell-role="th">3</td>
<td id="T_09f8c_row3_col0" class="data row3 col0">azido</td>
<td id="T_09f8c_row3_col1" class="data row3 col1">72</td>
</tr>
<tr class="odd">
<td id="T_09f8c_level0_row4" class="row_heading level0 row4" data-quarto-table-cell-role="th">4</td>
<td id="T_09f8c_row4_col0" class="data row4 col0">biotin</td>
<td id="T_09f8c_row4_col1" class="data row4 col1">72</td>
</tr>
</tbody>
</table>
&nbsp;&nbsp;&nbsp;
<table id="T_9ac4d" data-quarto-postprocess="true" class="table table-sm table-striped small">
<caption>Organism</caption>
<thead>
<tr class="header">
<th class="blank level0" data-quarto-table-cell-role="th">&nbsp;</th>
<th id="T_9ac4d_level0_col0" class="col_heading level0 col0" data-quarto-table-cell-role="th">org</th>
<th id="T_9ac4d_level0_col1" class="col_heading level0 col1" data-quarto-table-cell-role="th">count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td id="T_9ac4d_level0_row0" class="row_heading level0 row0" data-quarto-table-cell-role="th">0</td>
<td id="T_9ac4d_row0_col0" class="data row0 col0">Actinomycetospora chiangmaiensis</td>
<td id="T_9ac4d_row0_col1" class="data row0 col1">14</td>
</tr>
<tr class="even">
<td id="T_9ac4d_level0_row1" class="row_heading level0 row1" data-quarto-table-cell-role="th">1</td>
<td id="T_9ac4d_row1_col0" class="data row1 col0">Actinoplanes atraurantiacus</td>
<td id="T_9ac4d_row1_col1" class="data row1 col1">14</td>
</tr>
<tr class="odd">
<td id="T_9ac4d_level0_row2" class="row_heading level0 row2" data-quarto-table-cell-role="th">2</td>
<td id="T_9ac4d_row2_col0" class="data row2 col0">Paenibacillus polysaccharolyticus</td>
<td id="T_9ac4d_row2_col1" class="data row2 col1">14</td>
</tr>
<tr class="even">
<td id="T_9ac4d_level0_row3" class="row_heading level0 row3" data-quarto-table-cell-role="th">3</td>
<td id="T_9ac4d_row3_col0" class="data row3 col0">Okibacterium fritillariae</td>
<td id="T_9ac4d_row3_col1" class="data row3 col1">14</td>
</tr>
<tr class="odd">
<td id="T_9ac4d_level0_row4" class="row_heading level0 row4" data-quarto-table-cell-role="th">4</td>
<td id="T_9ac4d_row4_col0" class="data row4 col0">Nonomuraea candida</td>
<td id="T_9ac4d_row4_col1" class="data row4 col1">14</td>
</tr>
</tbody>
</table>

&nbsp;&nbsp;&nbsp;
</div>
</div>
<p>We have been able to reproduce most of the Robinson dataset, yielding measurements for 14 out of 15 compounds across 72 out of 73 organisms.</p>
</section>
<section id="exploring-the-data" class="level3">
<h3 class="anchored" data-anchor-id="exploring-the-data">Exploring the data <a class="anchor" id="1_3"></a></h3>
<p>We can investigate the data briefly before proceeding with PoET.</p>
<p>First, let’s examine the sequences from each organism to get an idea of sequence diversity:</p>
<div class="cell" data-execution_count="5">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> Bio <span class="im">import</span> pairwise2</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> align_report(seq1, seq2):</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Calculate the similarity between two sequences using the globalxx alignment score"""</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> seq1<span class="op">==</span>seq2:</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="dv">1</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    alignments <span class="op">=</span> pairwise2.align.globalxx(seq1, seq2)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    top_alignment <span class="op">=</span> alignments[<span class="dv">0</span>]</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(pairwise2.format_alignment(<span class="op">*</span>top_alignment) )</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> pairwise2.format_alignment(<span class="op">*</span>top_alignment)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> calculate_similarity(seq1, seq2):</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Calculate the similarity between two sequences using the globalxx alignment score"""</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> seq1<span class="op">==</span>seq2:</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="dv">1</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>    alignments <span class="op">=</span> pairwise2.align.globalxx(seq1, seq2)</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>    top_alignment <span class="op">=</span> alignments[<span class="dv">0</span>]</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>    alignment_score <span class="op">=</span> top_alignment[<span class="op">-</span><span class="dv">3</span>]</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>    similarity <span class="op">=</span> alignment_score <span class="op">/</span> <span class="bu">len</span>(seq1)</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> similarity</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_similarity_matrix(sequences):</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Generate a similarity matrix for a list of sequences"""</span></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>    num_sequences <span class="op">=</span> <span class="bu">len</span>(sequences)</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>    similarity_matrix <span class="op">=</span> np.zeros((num_sequences, num_sequences))</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(num_sequences):</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(i<span class="op">+</span><span class="dv">1</span>, num_sequences):</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>            similarity <span class="op">=</span> calculate_similarity(sequences[i], sequences[j])</span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>            similarity_matrix[i, j] <span class="op">=</span> similarity</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>            similarity_matrix[j, i] <span class="op">=</span> similarity</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>    np.fill_diagonal(similarity_matrix, <span class="dv">1</span>)</span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> similarity_matrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="6">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>unique_sequences <span class="op">=</span> data.drop_duplicates(<span class="st">'org'</span>).sequence.tolist()</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>unique_orgs <span class="op">=</span> data.drop_duplicates(<span class="st">'org'</span>).org.tolist()</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>similarity_matrix <span class="op">=</span> generate_similarity_matrix(unique_sequences)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>similarity_matrix <span class="op">=</span> pd.DataFrame(similarity_matrix, index<span class="op">=</span>unique_orgs, columns<span class="op">=</span>unique_orgs)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>sns.heatmap(similarity_matrix, cmap<span class="op">=</span><span class="st">'coolwarm'</span>, vmin<span class="op">=</span><span class="dv">0</span>, vmax<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Pairwise sequence Similarity"</span>)<span class="op">;</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-7-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="7">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Sequence similarity summary:"</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss">Mean: </span><span class="sc">{</span>np<span class="sc">.</span>mean(similarity_matrix.values[np.triu_indices(similarity_matrix.shape[<span class="dv">0</span>], k<span class="op">=</span><span class="dv">1</span>)])<span class="sc">:.2f}</span><span class="ss"> </span><span class="ch">\n\t</span><span class="ss">Min: </span><span class="sc">{</span>np<span class="sc">.</span><span class="bu">min</span>(similarity_matrix.values[np.triu_indices(similarity_matrix.shape[<span class="dv">0</span>], k<span class="op">=</span><span class="dv">1</span>)])<span class="sc">:.2f}</span><span class="ss"> </span><span class="ch">\n\t</span><span class="ss">Max: </span><span class="sc">{</span>np<span class="sc">.</span><span class="bu">max</span>(similarity_matrix.values[np.triu_indices(similarity_matrix.shape[<span class="dv">0</span>], k<span class="op">=</span><span class="dv">1</span>)])<span class="sc">:.2f}</span><span class="ss"> "</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Sequence similarity summary:
    Mean: 0.49 
    Min: 0.36 
    Max: 0.88 </code></pre>
</div>
</div>
<p>We can see above that there is a lot of diversity in our sequences; the average pairwise similarity is &lt;50%. And this covers a range from 36-88% (excluding the diagonal). This diversity of sequences makes the Robinson data very useful for evolutionary prompting.</p>
</section>
<section id="choosing-a-target-candidate" class="level3">
<h3 class="anchored" data-anchor-id="choosing-a-target-candidate">Choosing a target candidate<a class="anchor" id="1_4"></a></h3>
<p>We can now build our prompt, to do that we will first select a good compound for development (i.e.&nbsp;good data availability, broad range of activity and good specificity). Once we have selected our target compound we will take the most active enzyme sequence as a seed to generate an MSA, and then impose filters on that MSA to create a prompt.</p>
<p>First, let’s see how much non-zero data is available for each compound, choosing a compound with more data will make the resulting analysis more robust:</p>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compounds with most data</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>ordered_compounds <span class="op">=</span> data[data.mean_log_activity<span class="op">!=</span><span class="dv">0</span>].compound.value_counts().index.tolist()</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>data[data.mean_log_activity<span class="op">!=</span><span class="dv">0</span>].compound.value_counts()[::<span class="op">-</span><span class="dv">1</span>].plot(kind<span class="op">=</span><span class="st">'barh'</span>, figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">4</span>))<span class="op">;</span> </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Number of datapoints per compound"</span>)<span class="op">;</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Number of datapoints"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-9-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>We can see above the number of data points per compound; <code>azido</code>, <code>butoxy</code> and <code>heptynoate</code> have the most available data, covering &gt;90% of the enzymes.</p>
<p>We also want a target compound with a reasonable <em>range</em> of activities for two reasons: * 1) ideally measurement error should be smaller than true activity differences, this is more likely for larger data ranges * 2) a larger range suggests more scope for improved enzymatic activity, and a more interesting fitness landscape</p>
<p>I’ve plotted the range of activities below per compound:</p>
<div class="cell" data-execution_count="9">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>f, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">4</span>))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>sns.boxplot( data.query(<span class="st">"mean_log_activity!=0"</span>), x<span class="op">=</span><span class="st">'compound'</span>, y<span class="op">=</span><span class="st">'mean_log_activity'</span>,ax<span class="op">=</span>ax, order<span class="op">=</span>ordered_compounds)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">90</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Range of activity for each Compound"</span>)<span class="op">;</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Enzymatic activity (log10)"</span>)<span class="op">;</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">''</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-10-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>We can see <code>butoxy</code>, <code>hexanoate</code> and <code>heptanoate</code> all have broad ranges of activity. <code>Butoxy</code> was also noted above for its data availibility.</p>
<p>Lastly, let’s look at enzyme specificity, here defined as the correlation in activity across compounds. Specific enzymes are usually more highly valued than promiscuous enzymes, partly due to their greater efficiency and production of fewer by-products. Therefore we want to pick a compound that will enable us to prompt for specificity (as well as activity).</p>
<div class="cell" data-execution_count="10">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>compound_activity_corr <span class="op">=</span> np.zeros((<span class="bu">len</span>(ordered_compounds), <span class="bu">len</span>(ordered_compounds)))</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, compound1 <span class="kw">in</span> <span class="bu">enumerate</span>(ordered_compounds):</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j, compound2 <span class="kw">in</span> <span class="bu">enumerate</span>(ordered_compounds):</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>        compound1_data <span class="op">=</span> data[(data.compound<span class="op">==</span>compound1)]</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>        compound2_data <span class="op">=</span> data[(data.compound<span class="op">==</span>compound2)]</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>        compound_activity_corr[i, j] <span class="op">=</span> spearmanr(compound1_data.mean_log_activity, compound2_data.mean_log_activity)[<span class="dv">0</span>]</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>compound_activity_corr <span class="op">=</span> pd.DataFrame(compound_activity_corr, index<span class="op">=</span>ordered_compounds, columns<span class="op">=</span>ordered_compounds)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>f, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>sns.heatmap(compound_activity_corr, cmap<span class="op">=</span><span class="st">'YlOrRd'</span>, vmin<span class="op">=</span><span class="dv">0</span>, vmax<span class="op">=</span><span class="dv">1</span>, ax<span class="op">=</span>ax)<span class="op">;</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Spearman correlation of enzymatic activity per compound"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-11-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>This same data is summarized futher (mean correlation) in the table below:</p>
<div class="cell" data-execution_count="11">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>compound_activity_corr.mean(axis<span class="op">=</span><span class="dv">1</span>).<span class="bu">round</span>(<span class="dv">2</span>).reset_index().rename(columns<span class="op">=</span>{<span class="dv">0</span>:<span class="st">'mean correlation'</span>}).sort_values(<span class="st">'mean correlation'</span>, ascending<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="11">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">index</th>
<th data-quarto-table-cell-role="th">mean correlation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>heptanoate</td>
<td>0.60</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4</td>
<td>hexanoate</td>
<td>0.56</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>decanoate</td>
<td>0.53</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2</td>
<td>heptynoate</td>
<td>0.51</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">5</td>
<td>cyclopentyl</td>
<td>0.51</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>dodecanoate</td>
<td>0.51</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>7Ph_heptanoate</td>
<td>0.49</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">0</td>
<td>azido</td>
<td>0.47</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">7</td>
<td>oxadiazole</td>
<td>0.42</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>TMA</td>
<td>0.40</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1</td>
<td>butoxy</td>
<td>0.39</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">11</td>
<td>ClPh_propionate</td>
<td>0.39</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">12</td>
<td>biotin</td>
<td>0.39</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">13</td>
<td>dimethyl</td>
<td>0.35</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>We can see most specific enzymes target <code>dimethyl</code>, with <code>butoxy</code> a little further up the list.</p>
<p>Based on the above findings, <a href="https://pubchem.ncbi.nlm.nih.gov/compound/2-_2-Butoxyethoxy_ethyl-acetate#section=Molecular-Formula">butoxy</a> stands out as a promising compound to focus on for several reasons:</p>
<ul>
<li>Extensive experimental data: Butoxy has many non-zero data points.</li>
<li>Wide range activity: Butoxy activity spans 1-2 orders of magnitude.</li>
<li>Specificity: Butoxy is only weakly correlated to other compounds.</li>
</ul>
<p>Moreover, the most recent <a href="https://pubchem.ncbi.nlm.nih.gov/compound/2-_2-Butoxyethoxy_ethyl-acetate#section=U-S-Production">data</a> available lists production of butoxy at 10-50 million pounds per year in the USA. This is used in diverse applications such as:</p>
<ul>
<li>Paint additives: plasticizers in paints improve durability and flexibility.</li>
<li>Cosmetics and personal care: butoxy solvents prevent ingredient separation and maintain clarity, even at high temperatures.</li>
<li>Hygeine and cleaning: due to its effective ability to dissolve grease and dirt.</li>
</ul>
<p>Considering these factors, the development of an enhanced butoxy-specific thiolase would yield significant industrial benefits via potential gains in product performance, cost-effectiveness, and sustainability.</p>
<div class="cell" data-execution_count="12">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>selected_compound <span class="op">=</span> <span class="st">'butoxy'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Let’s briefly examine the <code>butoxy</code> data a little more closely.</p>
<p>Below is the correlation between <code>butoxy</code> activity and every other compound in our Robinson dataset. We can see the weakest correlation is against <code>dimethyl</code>, <code>ClPhpropionate</code> and <code>oxadiazole</code>. These will serve as interesting negative controls later on.</p>
<div class="cell" data-execution_count="13">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>compound_correlations <span class="op">=</span> compound_activity_corr[selected_compound].sort_values().<span class="bu">round</span>(<span class="dv">2</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>least_correlated_compound <span class="op">=</span> compound_correlations.index[<span class="dv">0</span>]</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>compound_correlations.reset_index().rename(columns<span class="op">=</span>{<span class="st">'index'</span>:<span class="st">'compound'</span>, selected_compound:<span class="st">'correlation'</span>} )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="13">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">compound</th>
<th data-quarto-table-cell-role="th">correlation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>dimethyl</td>
<td>0.12</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>ClPh_propionate</td>
<td>0.16</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>oxadiazole</td>
<td>0.25</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>biotin</td>
<td>0.25</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>TMA</td>
<td>0.30</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>decanoate</td>
<td>0.33</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>dodecanoate</td>
<td>0.36</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>heptanoate</td>
<td>0.40</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>cyclopentyl</td>
<td>0.40</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>hexanoate</td>
<td>0.42</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>azido</td>
<td>0.47</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">11</td>
<td>7Ph_heptanoate</td>
<td>0.48</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">12</td>
<td>heptynoate</td>
<td>0.55</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">13</td>
<td>butoxy</td>
<td>1.00</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>Let’s also take a closer look at the distribution of <code>butoxy</code> activity per enzyme (below), we can see this spans 1-2 orders of magnitude, with K. sedentarius having the most active thiolase:</p>
<div class="cell" data-execution_count="14">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>f, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">4</span>))</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>sns.barplot( data.query(<span class="st">"compound== @selected_compound "</span>).sort_values(<span class="st">'mean_log_activity'</span>), x<span class="op">=</span><span class="st">'org'</span>, y<span class="op">=</span><span class="st">'mean_log_activity'</span>,ax<span class="op">=</span>ax, color<span class="op">=</span><span class="st">'darkblue'</span> )</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">90</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f"Range of activity for </span><span class="sc">{</span>selected_compound<span class="sc">}</span><span class="ss">"</span>)<span class="op">;</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Enzymatic activity (log10)"</span>)<span class="op">;</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">''</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-15-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="summary" class="level3">
<h3 class="anchored" data-anchor-id="summary">Summary <a class="anchor" id="1_5"></a></h3>
<p>We have 72 enzymes that catalyse 14 thiols with a range of activity. We have selected <code>butoxy</code> as our target compound due to it’s data availability and industrial importance. <em>Kytococcus sedentarius</em> represents the most active butoxy thiolase in our dataset.</p>
</section>
</section>
<section id="poet-thiolase-scoring" class="level2">
<h2 class="anchored" data-anchor-id="poet-thiolase-scoring">PoET thiolase scoring <a class="anchor" id="2_0"></a></h2>
<section id="job-creation" class="level3">
<h3 class="anchored" data-anchor-id="job-creation">Job creation <a class="anchor" id="2_1"></a></h3>
<p>Having explored the data and selected <code>butoxy</code> catalysis as our main goal, we can now proceed to the PoET job, and scoring all 72 enzymes.</p>
<p>We will build a prompt around the most active <code>butoxy</code> catalysing enzyme (from <code>K. sedentarius</code>), and compare our model scores with experimental data from Robinson.</p>
<div class="cell" data-execution_count="15">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> data.query(<span class="st">"compound==@selected_compound"</span>).sort_values(<span class="st">'mean_log_activity'</span>, ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>msa_seed <span class="op">=</span> x.sequence.tolist()[<span class="dv">0</span>]</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>msa_seed_id <span class="op">=</span> x.org.tolist()[<span class="dv">0</span>] </span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Selected compound: </span><span class="sc">{</span>selected_compound<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Most active </span><span class="sc">{</span>selected_compound<span class="sc">}</span><span class="ss"> enzyme: </span><span class="sc">{</span>msa_seed_id<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Enzyme sequence:"</span>, msa_seed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Selected compound: butoxy
Most active butoxy enzyme: Kytococcus sedentarius
Enzyme sequence: MPGNATFRHSNTSVLSVTALPAPEVHTSAEFDARIMDTLRRNKLVPGVLSRLVGIKERRWWPEDVHFTEAAAEAGRRALVDAGIKPEQVGLMVNTSVSRDHLEPSTAVRIHDTLGLPRSASNFDVTNACLGFVNGMQIAAAMIDSGQIEYALIVNAETTRHTHETTLKRLESEDATADDVLQQLATLTLGSGAAAMVLGRTDKHPEGHRFVGGVSRAGTEHHELCVGDFDRMTTDSVGLQEAGIALSEELWAEAAEEFDWQHLDRYVIHQVSTVHTEQICDRLGLDGAKVPRTFPTYGNMGPAAVPFTLALEAPHLERGDRVLLMGIGSGLNACCAEIEW</code></pre>
</div>
</div>
<div class="cell" data-execution_count="16">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># one sequence per organism, save to file</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>data.query(<span class="st">"compound==@selected_compound"</span>)[<span class="st">'sequence'</span>].to_csv(<span class="st">"./data/query_data.csv"</span>, index<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We can now kick off the job!</p>
<p>I will upload the query sequences (to be scored) as well as the seed sequence (to generate the MSA and prompt). For the prompt, I have selected hyper-parameters that will filter the MSA down to sequences that are at least somewhat similar to the seed sequence.</p>
<p>The exact arguments are: * Use first sequence as seed to generate MSA = <code>True</code> (i.e.&nbsp;use <em>K. sedentarius</em> sequence to generate MSA) * number of prompts to ensemble = <code>1</code> (ensemble only 1 replicate for now) * minimum similarity to seed = <code>0.5</code> (ensure all prompt sequences have &gt;50% sequence similarity to the seed. This is important to select for butoxy activity specifically, and not just thiol activity in general) * Homology level = <code>0.5</code> (the tolerance for redundant sequences in the prompt, sequences with &gt;0.5 similarity will be considered redundant and downsampled) * always include seed sequence: <code>True</code> (i.e.&nbsp;ensure <em>K. sedentarius</em> sequence is always included in the prompt) * maximum number of sequences = <code>50</code> (the total number of sequences per replicate) * random seed = <code>42</code> (for reproducible results)</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/356e8cdf-1-robinson_score.gif" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">robinson_score.gif</figcaption>
</figure>
</div>
<div class="cell" data-execution_count="17">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>results_path <span class="op">=</span> <span class="st">"./results/Robinson2020/prots2prot-query.csv"</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>scoring_job_id <span class="op">=</span> <span class="st">'26a43427-796e-4fc6-8c4a-31ca6109fa0b'</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>evo_prompt_id <span class="op">=</span> <span class="st">'3e581bc1-353e-4dfc-aeb4-82e1e6b5843e'</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>scoring_result <span class="op">=</span> pd.read_csv(results_path)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Reading from </span><span class="sc">{</span>scoring_job_id<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>scoring_result.head(<span class="dv">3</span>)</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Reading from 26a43427-796e-4fc6-8c4a-31ca6109fa0b</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="17">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">scores</th>
<th data-quarto-table-cell-role="th">sequence</th>
<th data-quarto-table-cell-role="th">score</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>[-492.5433044433594]</td>
<td>MRAAQRYDDVVITGLAHVDAPHVVTSTELEDRLAGTLGRLRITPGL...</td>
<td>-492.543304</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>[-703.46240234375]</td>
<td>MNVGITGIAYALPAREVTTAELQRRVAPKLPPRMFEQATGITRRRV...</td>
<td>-703.462402</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>[-219.8573303222656]</td>
<td>MDGNATTTHRNVALLSVATTMATRITTSAEIDRRLAPSLKRLKLPT...</td>
<td>-219.857330</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="prompting-poet" class="level3">
<h3 class="anchored" data-anchor-id="prompting-poet">Prompting PoET <a class="anchor" id="2_2"></a></h3>
<p>Let’s examine the MSA constructed from our seed sequence (before any filtering, such as <code>minimum similarity</code> is applied):</p>
<div class="cell" data-execution_count="18">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>response <span class="op">=</span> session.prots2prot.get_msa(scoring_job_id)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>msadf <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(response), columns<span class="op">=</span>[<span class="st">'name'</span>,<span class="st">'sequence'</span>])</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Number of sequences in MSA:"</span>, <span class="bu">len</span>(msadf))</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Sample:"</span>)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>msadf.head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Number of sequences in MSA: 1604
Sample:</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="18">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>seed</td>
<td>MPGNATFRHSNTSVLSVTALPAPEVHTSAEFDARIMDTLRRNKLVP...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>A0A892J436</td>
<td>MPGNATFRHTNTSVLSVTGLPAPEVHTSAEFDARIMDTLRRNKLVP...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>A0A1F1KA82</td>
<td>MPGNATFRHSNTSILSVTGLPAPEVRTSAEFDERIMDTLRRNRLVP...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>A0A4Q7V0T1</td>
<td>MTGTALFHHDNTVIVSVCGIDAPTVVTSAEIDDRLSDTYERLELRP...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>A0A516PWD2</td>
<td>MSGNATYRLTNTSMLSVSAIEAPVVATSEEFDDRLEATYRRTGLRR...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>A0A495XTP4</td>
<td>MSGNATFRYANTAVLSVCAVDAPVVMTSDAFDEKLLETYARVGLRS...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>A0A839QRV3</td>
<td>MGGNATFRHDNSALLAIGSVEAPVVVTSDEFDRRLAPSLKRLRLSR...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>A0A176UIU6</td>
<td>MIGNATFRHRNTALLSVISVQAPRIVSSTEFDHRLASTLRRLKFPP...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>A0A838H8R0</td>
<td>MTGNGIFRSSNTSVLAVDALDAPRVVTSAELDERLSATYRRVGLRP...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>A0A3N0C1T5</td>
<td>MIGNATFRHRNTALLSVSSVEAPKVVSSAEFDHMLAPTLQRLKFPP...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>We have &gt;1600 sequences in the MSA, and we told the prompt to take up to 50 (per replicate).</p>
<p>Let’s see how many sequences are in the prompt:</p>
<div class="cell" data-execution_count="19">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>response <span class="op">=</span> session.prots2prot.get_prompt(scoring_job_id, prompt_index<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>promptdf <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(response), columns<span class="op">=</span>[<span class="st">'name'</span>,<span class="st">'sequence'</span>])</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">#rename seed from 101 to uniprotkb entry </span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>promptdf.loc[promptdf.name<span class="op">==</span><span class="st">'101'</span>, <span class="st">'name'</span>] <span class="op">=</span> <span class="st">'C7NEV1'</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Number of sequences in prompt:"</span>, <span class="bu">len</span>(promptdf))</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Sample:"</span>)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>promptdf.head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Number of sequences in prompt: 50
Sample:</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="19">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>seed</td>
<td>MPGNATFRHSNTSVLSVTALPAPEVHTSAEFDARIMDTLRRNKLVP...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>A0A1F1KA82</td>
<td>MPGNATFRHSNTSILSVTGLPAPEVRTSAEFDERIMDTLRRNRLVP...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>A0A516PWD2</td>
<td>MSGNATYRLTNTSMLSVSAIEAPVVATSEEFDDRLEATYRRTGLRR...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>A0A543KKX1</td>
<td>MNGNATFRHHDTAVLSVTAVDAPVVKTSDDFDEVLADSYRRNGLRP...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>A0A098YAU2</td>
<td>MSGNATHRFSNTSILTVQTADASQVVTSDDIDQRLADTYKRVGLRP...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>UPI001C43CB83</td>
<td>MPGNATFHHRNASLLSVTEVQAPEVLTSKDLDKRLGDALKRLRLPT...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>UPI000975AA8B</td>
<td>MPGNATTHFDDVALVSVASVLPSRVTTSDDIEERLAPALKRLKLKP...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>UPI001C67085F</td>
<td>MSGNITFRHHDTAVLAVTAVDAPVVKTSEEFDAIIGDSYTRNGLRP...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>UPI00047A1E2B</td>
<td>MTGNFNFRHRNTALLGVESVEAPVIITSAEFDRILSPALRRLRLSK...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>UPI001963D3D9</td>
<td>MAGNAIHRFANTTVLAVCAIDAPRVVTSAEIDERLADVYTRVSLRP...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>The prompt filtering resulted in 50 sequences, including the seed sequence (<code>C7NEV1</code>). Exactly as expected (though &lt;50 would also be expected with more stringent filtering).</p>
<p>We can also sense-check that the prompt doesn’t simply contain the query sequences:</p>
<div class="cell" data-execution_count="20">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use all the prompts combined</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>response <span class="op">=</span> session.prots2prot.get_prompt(scoring_job_id)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>promptdf <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(response), columns<span class="op">=</span>[<span class="st">'name'</span>,<span class="st">'sequence'</span>])</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>promptdf <span class="op">=</span> promptdf.drop_duplicates().reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Total sequences across all replicates:"</span>, promptdf.shape[<span class="dv">0</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Total sequences across all replicates: 50</code></pre>
</div>
</div>
<div class="cell" data-execution_count="21">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Number of overlapping sequences in prompt and query: </span><span class="sc">{</span><span class="bu">len</span>(<span class="bu">set</span>(unique_sequences) <span class="op">&amp;</span> <span class="bu">set</span>(promptdf.sequence.unique().tolist()))<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> [i <span class="cf">for</span> i <span class="kw">in</span> unique_sequences <span class="cf">if</span> i <span class="kw">in</span> promptdf.sequence.unique().tolist()]</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>data.query(<span class="st">"compound==@selected_compound"</span>).set_index(<span class="st">'sequence'</span>).loc[ x ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Number of overlapping sequences in prompt and query: 2</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="21">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">org</th>
<th data-quarto-table-cell-role="th">compound</th>
<th data-quarto-table-cell-role="th">mean_log_activity</th>
<th data-quarto-table-cell-role="th">std_log_activity</th>
<th data-quarto-table-cell-role="th">accession</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sequence</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">MPGNATFRHSNTSVLSVTALPAPEVHTSAEFDARIMDTLRRNKLVPGVLSRLVGIKERRWWPEDVHFTEAAAEAGRRALVDAGIKPEQVGLMVNTSVSRDHLEPSTAVRIHDTLGLPRSASNFDVTNACLGFVNGMQIAAAMIDSGQIEYALIVNAETTRHTHETTLKRLESEDATADDVLQQLATLTLGSGAAAMVLGRTDKHPEGHRFVGGVSRAGTEHHELCVGDFDRMTTDSVGLQEAGIALSEELWAEAAEEFDWQHLDRYVIHQVSTVHTEQICDRLGLDGAKVPRTFPTYGNMGPAAVPFTLALEAPHLERGDRVLLMGIGSGLNACCAEIEW</td>
<td>Kytococcus sedentarius</td>
<td>butoxy</td>
<td>2.968362</td>
<td>0.087611</td>
<td>WP_012802190.1_Kytococcus_sedentarius</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">MSENASYRLANCSILSVSTVEAPEVVTSDHLDELLQPAYDTTGMMAGQIERLAGVRERRWFPKDADYTRGAVEAGRQAIADARIDADRIGLIVNASVTRPHLEPGISAMVHAELELPRSCIAFDVTNACLGVVNSLQIAGTMIDAGQIEYALIVASEGARQMQEASIARLVDGGSRQDVKDAFATMTLGSGAVGIVLGRSDRHPEGHQIIGGITRAGTEHHELCIGGMDGMRTDASKLFVEGLSLATDAWAEAKGEWDWEEMDWYVAHQTSTAHIACLCDTLKLPAEKFPVTVDTYGNIGPVALPFTLGMYQHKMRKGDRVLLMGIGSGLNTSMTEVRW</td>
<td>Mobilicoccus massiliensis</td>
<td>butoxy</td>
<td>2.360310</td>
<td>0.106454</td>
<td>WP_040156198.1_Mobilicoccus_massiliensis</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div class="cell" data-execution_count="22">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Sequence similarity K. sedentarius v  M. massiliensis: </span><span class="sc">{</span>calculate_similarity(x[<span class="dv">0</span>], x[<span class="dv">1</span>])<span class="sc">:.2f}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Sequence similarity K. sedentarius v  M. massiliensis: 0.56</code></pre>
</div>
</div>
<p>As well containing the original seed (as requested in our prompt definition), the prompt also contains a sequence from <em>Mobilicoccus massiliensis</em> (which is also part of the <em>Micrococcales</em> order, and also highly active against butoxy - so no surprise our MSA captured it).</p>
<p>The remaining 48 sequences are from organisms not in our dataset. Let’s try to find these in Uniprot. As the MSA uses ColabFold, which contains metagenomic and transcriptomic data, we may not find all the sequences in the UniProtKB. These will be marked <code>notfound</code> below:</p>
<div class="cell" data-execution_count="23">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>prompt_orgs <span class="op">=</span> [] </span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> accnumber <span class="kw">in</span> promptdf.name.unique()[<span class="dv">0</span>:]:</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> accnumber <span class="op">==</span> <span class="st">'101'</span>:</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>         <span class="cf">continue</span> </span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>    r <span class="op">=</span> requests.get(<span class="ss">f"https://www.uniprot.org/uniprotkb/</span><span class="sc">{</span>accnumber<span class="sc">}</span><span class="ss">.json"</span>, timeout<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>    prompt_orgs.append((accnumber, r.json().get(<span class="st">'organism'</span>, {}).get(<span class="st">'scientificName'</span>, <span class="st">'notfound'</span>), r.json().get(<span class="st">'organism'</span>, {}).get(<span class="st">'lineage'</span>, [] )))</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>promptdf <span class="op">=</span> pd.DataFrame(prompt_orgs, columns<span class="op">=</span>[<span class="st">'accnumber'</span>, <span class="st">'organism'</span>, <span class="st">'lineage'</span>]).merge(promptdf, left_on<span class="op">=</span><span class="st">'accnumber'</span>, right_on<span class="op">=</span><span class="st">'name'</span>)</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="co"># we know M. massiliensis </span></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> promptdf.loc[promptdf.accnumber<span class="op">==</span><span class="st">'UPI00059046A7'</span>, <span class="st">'sequence'</span>].tolist()[<span class="dv">0</span>]<span class="op">==</span> data.loc[data.org<span class="op">==</span><span class="st">'Mobilicoccus massiliensis'</span>, <span class="st">'sequence'</span>].tolist()[<span class="dv">0</span>]</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>promptdf.loc[promptdf.accnumber<span class="op">==</span><span class="st">'UPI00059046A7'</span>, <span class="st">'organism'</span>] <span class="op">=</span> <span class="st">'Mobilicoccus massiliensis'</span></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>promptdf.organism.value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="23">
<pre><code>organism
notfound                                                                            24
Plantibacter sp. VKM Ac-1784                                                         1
Pirellulimonas nuda                                                                  1
Brevibacterium luteolum                                                              1
Lysobacter sp. Root604                                                               1
Xanthomonas sacchari                                                                 1
Lysobacter sp                                                                        1
Nakamurella sp. PAMC28650                                                            1
Acidimicrobiia bacterium                                                             1
hydrothermal vent metagenome                                                         1
Microbacterium halimionae                                                            1
Egicoccus halophilus                                                                 1
Quadrisphaera granulorum                                                             1
bacterium (Candidatus Blackallbacteria) CG17_big_fil_post_rev_8_21_14_2_50_48_46     1
Kytococcus sp. HMSC28H12                                                             1
Actinomycetia bacterium                                                              1
Microlunatus antarcticus                                                             1
Microlunatus flavus                                                                  1
Mobilicoccus massiliensis                                                            1
Pseudoclavibacter sp. RFBA6                                                          1
Tessaracoccus antarcticus                                                            1
Geodermatophilus sp. LHW52908                                                        1
Georgenia soli                                                                       1
Modestobacter caceresii                                                              1
Ornithinimicrobium humiphilum                                                        1
Microlunatus elymi                                                                   1
Wenzhouxiangella sp. W260                                                            1
Name: count, dtype: int64</code></pre>
</div>
</div>
<p>We can see above we were able to ascertain the origins of just over half the prompt sequences, suggesting the other half came from metagenomic or transcriptomic sources.</p>
<p>Below, I have summarised the taxonomic order for the sequences present in UniProtKB:</p>
<div class="cell" data-execution_count="24">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>promptdf.loc[promptdf[<span class="st">'lineage'</span>].<span class="bu">apply</span>(<span class="bu">len</span>) <span class="op">&gt;=</span><span class="dv">3</span>].lineage.<span class="bu">apply</span>(<span class="kw">lambda</span> x: x[<span class="dv">2</span>]).value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="24">
<pre><code>lineage
Micrococcales             7
Propionibacteriales       4
Gammaproteobacteria       4
Geodermatophilales        2
Kineosporiales            1
Nitriliruptoria           1
ecological metagenomes    1
Acidimicrobiia            1
Nakamurellales            1
Planctomycetia            1
Name: count, dtype: int64</code></pre>
</div>
</div>
<p>We can see the prompt is quite diverse, containing 10 different taxonomic orders.</p>
</section>
<section id="scoring-results" class="level3">
<h3 class="anchored" data-anchor-id="scoring-results">Scoring results <a class="anchor" id="2_3"></a></h3>
<p>We can now turn to the PoET results. PoET modelling will return a log-liklihood (LL) score to evaluate the fitness of our queries against the prompt. The higher the score (i.e.&nbsp;less negative) the better the sequence (the more fit).</p>
<p>We can download our results for further analysis from the UI:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/8dc3a16b-1-robinson_getscore.gif" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">robinson_getscore.gif</figcaption>
</figure>
</div>
<div class="cell" data-execution_count="25">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>scoring_result.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="25">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">scores</th>
<th data-quarto-table-cell-role="th">sequence</th>
<th data-quarto-table-cell-role="th">score</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>[-492.5433044433594]</td>
<td>MRAAQRYDDVVITGLAHVDAPHVVTSTELEDRLAGTLGRLRITPGL...</td>
<td>-492.543304</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>[-703.46240234375]</td>
<td>MNVGITGIAYALPAREVTTAELQRRVAPKLPPRMFEQATGITRRRV...</td>
<td>-703.462402</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>[-219.8573303222656]</td>
<td>MDGNATTTHRNVALLSVATTMATRITTSAEIDRRLAPSLKRLKLPT...</td>
<td>-219.857330</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>[-328.9368896484375]</td>
<td>MRGNATTRHSNVALLGITEAVAPVEVSSKSFDDRLADTLKTLRLPK...</td>
<td>-328.936890</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>[-766.7886352539062]</td>
<td>MTYSRIGQIAVHLPEGGRTSAEVEALVAELNPGFEPPEGLIERFTG...</td>
<td>-766.788635</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>We expect 72 results, one for each enzyme uploaded:</p>
<div class="cell" data-execution_count="26">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(scoring_result.shape[<span class="dv">0</span>], <span class="st">"results found"</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>compound_scores <span class="op">=</span> scoring_result.merge(data.query(<span class="st">"compound==@selected_compound"</span>), on<span class="op">=</span><span class="st">'sequence'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>72 results found</code></pre>
</div>
</div>
<p>Now let’s see how well our PoET model scores correlate with the experimental activity data against <code>butoxy</code>:</p>
<div class="cell" data-execution_count="27">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> compound_scores.query(<span class="st">"mean_log_activity&gt;0"</span>) <span class="co"># drop 0 active rows</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>r,p <span class="op">=</span> spearmanr(x[<span class="st">'score'</span>], x[<span class="st">'mean_log_activity'</span>])</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>r,p <span class="op">=</span> np.<span class="bu">round</span>(r,<span class="dv">2</span>), np.<span class="bu">round</span>(p,<span class="dv">2</span>)</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Spearman correlation between score and mean_log_activity: </span><span class="sc">{</span>r<span class="sc">:.2f}</span><span class="ss">, p=</span><span class="sc">{</span>p<span class="sc">:.2f}</span><span class="ss">"</span>)</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>f, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">12</span>,<span class="dv">6</span>), sharex<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> axs</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="co"># full results</span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>sns.regplot(data<span class="op">=</span>x,</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>            x<span class="op">=</span><span class="st">'mean_log_activity'</span>,</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>            y<span class="op">=</span><span class="st">'score'</span>,</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>            scatter_kws<span class="op">=</span>{<span class="st">'s'</span>: <span class="dv">5</span>, <span class="st">'alpha'</span>: <span class="fl">0.5</span>},</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>            n_boot<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>            robust<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>            ax<span class="op">=</span>ax )</span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>ax .set_xlabel(<span class="st">"Experimental "</span>)</span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a>ax .set_ylabel(<span class="st">"Model score"</span>)</span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a><span class="co"># add spearmanr r and p results to graph</span></span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a>ax .text(<span class="fl">0.05</span>, <span class="fl">0.9</span>, <span class="ss">f"spearman r=</span><span class="sc">{</span>r<span class="sc">:.2f}</span><span class="ch">\n</span><span class="ss">p-value =</span><span class="sc">{</span>p<span class="sc">:.2f}</span><span class="ss">"</span>, transform<span class="op">=</span>plt.gca().transAxes, size<span class="op">=</span><span class="dv">14</span>)<span class="op">;</span></span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a>ax.set_title(<span class="ss">f"Correlation between </span><span class="sc">{</span>selected_compound<span class="sc">}</span><span class="ss"> and log-liklihood score</span><span class="ch">\n</span><span class="ss"> MSA seed: </span><span class="sc">{</span>msa_seed_id<span class="sc">}</span><span class="ss">"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Spearman correlation between score and mean_log_activity: 0.41, p=0.00</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-28-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>Excellent, a &gt;0.4 correlation (p&lt;0.05) between activity and model score with just a single sequence, and no wetlab data!</p>
<p>We can also examine how it performed against compounds that were weakly correlated with <code>butoxy</code> to check specificity (see the reminder below) .</p>
<div class="cell" data-execution_count="28">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Weakest correlation to butoxy activity:"</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>compound_correlations.head(<span class="dv">3</span>).reset_index().rename( columns<span class="op">=</span>{<span class="st">'index'</span>:<span class="st">'compound'</span>, selected_compound:<span class="st">'correlation'</span>} )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Weakest correlation to butoxy activity:</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="28">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">compound</th>
<th data-quarto-table-cell-role="th">correlation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>dimethyl</td>
<td>0.12</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>ClPh_propionate</td>
<td>0.16</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>oxadiazole</td>
<td>0.25</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>Given the nature of the prompt we supplied, we expect there to be very little correlation between our model scores and experimental activity for these 3 compounds:</p>
<div class="cell" data-execution_count="29">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>f, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">3</span>,figsize<span class="op">=</span>(<span class="dv">15</span>,<span class="dv">5</span>), sharex<span class="op">=</span><span class="va">True</span>, sharey<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>axs <span class="op">=</span> axs.flatten()</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, compound <span class="kw">in</span>  <span class="bu">enumerate</span>(compound_correlations.index.tolist()[<span class="dv">0</span>:<span class="dv">3</span>]):</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> scoring_result.merge(data.query(<span class="st">"compound==@compound"</span>), on<span class="op">=</span><span class="st">'sequence'</span>)</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> x.query(<span class="st">"mean_log_activity!=0"</span>)</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>    r,p <span class="op">=</span> spearmanr(x[<span class="st">'score'</span>], x[<span class="st">'mean_log_activity'</span>])  </span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>    sns.regplot(data<span class="op">=</span>x,</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>                x<span class="op">=</span><span class="st">'mean_log_activity'</span>,</span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>                y<span class="op">=</span><span class="st">'score'</span>,</span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>                scatter_kws<span class="op">=</span>{<span class="st">'s'</span>: <span class="dv">5</span>, <span class="st">'alpha'</span>: <span class="fl">0.5</span>},</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>                n_boot<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a>                robust<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a>                ax<span class="op">=</span>axs[i] )</span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>    axs[i].set_ylabel(<span class="st">"model score"</span>)</span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a>    axs[i].set_xlabel(<span class="st">"Activity (log10)"</span>)</span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add spearmanr r and p results to graph</span></span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a>    axs[i].set_title(<span class="ss">f"</span><span class="sc">{</span>compound<span class="sc">}</span><span class="ss"> vs score </span><span class="sc">{</span>r<span class="op">=</span><span class="sc">:.2f}</span><span class="ss">, </span><span class="sc">{</span>p<span class="op">=</span><span class="sc">:.2f}</span><span class="ss">"</span>)<span class="op">;</span></span>
<span id="cb43-19"><a href="#cb43-19" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/opt/conda/envs/py38/lib/python3.8/site-packages/statsmodels/robust/robust_linear_model.py:288: ConvergenceWarning: Estimated scale is 0.0 indicating that the most last iteration produced a perfect fit of the weighted data.
  warnings.warn('Estimated scale is 0.0 indicating that the most'</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-30-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>As expected, our p-values are all &gt;0.05, indicating a non-significant relationship between fitness (which we defined using a butoxy thiolase) and experimental activity against <code>dimethyl</code>, <code>ClPh propionate</code> and <code>oxadiazole</code>! This suggests our prompt was succesful in encoding the features that make our butoxy thiolase specific and active!</p>
<p>Note that the key to success here is two things (in order of importance): * 1) <code>minimum similarity</code> sets a floor in how similar our prompt sequences are to the <em>K. sedentarius</em> seed. This is important as we need our prompt to reflect butoxy activity specifically (and not thiolases in general). * 2) <code>homology level</code> sets a ceiling on redundant sequences, ensuring the prompt is still diverse even after the minimum similarity step.</p>
<p>For reference, you can see the results from the default prompt settings (<em>min. similarity = 0, homology level =0.8</em>) in the <a href="#defaultprompt">appendix</a>. This yields a correlation coefficient of 0.26 (p=0.03). Which is significant, but far below the 0.4 with the more thoughtful prompt above.</p>
</section>
<section id="poet-scoring-summary" class="level3">
<h3 class="anchored" data-anchor-id="poet-scoring-summary">PoET scoring Summary <a class="anchor" id="2_4"></a></h3>
<p>This is a data-free way to start interrogating sequences for fitness, in this example, we took a series of thiolases and created a prompt to reflect butoxy catalysing activity. We achieved this by using the most active enzyme for butoxy catalysis as a seed for an MSA, and then a series of filters to ensure the resulting prompt looked reasonably like the original enzyme.</p>
<p>We then tested the results of this approach across 72 thiolases from 10 different taxonomic classes. This resulted in a strong (&gt;0.4) spearman correlation between PoET model scores and experimental activity. We confirmed that this was a specific effect by comparing those same scores to <code>dimethyl</code> (and other weakly correlated thiols) catalysis - and found no significant relationship (as expected).</p>
</section>
</section>
<section id="poet-de-novo-thiolase-generation" class="level2">
<h2 class="anchored" data-anchor-id="poet-de-novo-thiolase-generation">PoET de novo thiolase generation <a class="anchor" id="3_0"></a></h2>
<section id="job-creation-1" class="level3">
<h3 class="anchored" data-anchor-id="job-creation-1">Job creation <a class="anchor" id="3_1"></a></h3>
<p>Now that we have a meaningful prompt, we can re-use it to generate entirely new sequences. We can use these new sequences to understand <code>butoxy</code> thiolases, and to explore sequence and structural constraints in <code>butoxy</code> catalysis. Additionally, we could run these through wetlab assays to generate more data, and start exploring the fitness landscape with empirical observations. For now, we will focus on computational insights.</p>
<p>To kick off the PoET generation job we will supply a few key arguments: * Number of sequences to generate = <code>50</code> * top-k = <code>10</code> (at each step, sample from only the 10 most likely amino acids). * Maximum generated protein length = <code>500</code> (i.e.&nbsp;stop generating any protein beyond this size)</p>
<p>We have left the following arguments at their default value: * temperature = <code>1.0</code> (this will control the creativity of the model, higher temperatures lead to more diverse output. 1.0 is a good default value) * top-p = <code>1.0</code> (this is an alternative to top-k, where we sample from the amino acids within cumulative <em>p</em> probability distribution)</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/687fe700-1-image.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">image.png</figcaption>
</figure>
</div>
<div class="cell" data-execution_count="30">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>gen_results_path <span class="op">=</span> <span class="st">"./results/Robinson2020/Robinson_generate.csv"</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>gen_job_id <span class="op">=</span> <span class="st">'5597d678-0bc9-4828-90af-2527e6898977'</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Reading from job </span><span class="sc">{</span>gen_job_id<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>gen_score_result <span class="op">=</span> pd.read_csv(gen_results_path)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>gen_score_result <span class="op">=</span> gen_score_result.sort_values(by<span class="op">=</span><span class="st">'score'</span>, ascending<span class="op">=</span><span class="va">False</span>).reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>gen_score_result[<span class="st">'name'</span>] <span class="op">=</span> [<span class="ss">f"gen-</span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">"</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(gen_score_result.shape[<span class="dv">0</span>])]</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>gen_score_result[[<span class="st">'name'</span>,<span class="st">'sequence'</span>]].to_csv(<span class="st">"./results/Robinson2020/colabfold_input.csv"</span>, index<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Reading from job 5597d678-0bc9-4828-90af-2527e6898977</code></pre>
</div>
</div>
</section>
<section id="evaluating-generated-sequences" class="level3">
<h3 class="anchored" data-anchor-id="evaluating-generated-sequences">Evaluating generated sequences <a class="anchor" id="3_2"></a></h3>
<p>We have generated a variety of sequences with different scores, let’s first download the novel sequences and scores like we did previously:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/9bf759f4-1-robinson2_genresults.gif" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">robinson2_genresults.gif</figcaption>
</figure>
</div>
<div class="cell" data-execution_count="31">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>gen_score_result.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="31">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">scores</th>
<th data-quarto-table-cell-role="th">sequence</th>
<th data-quarto-table-cell-role="th">score</th>
<th data-quarto-table-cell-role="th">name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>[-105.93228149414062]</td>
<td>MSENATYRLANTSVLSVSTIEAPEVVTSDHLDELLQPAYDSTGMMA...</td>
<td>-105.932281</td>
<td>gen-1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>[-114.87188720703125]</td>
<td>MGRNATHRFGNTSVLSVCAIEAPRTITSAEIDERLAPVYQRVGLRA...</td>
<td>-114.871887</td>
<td>gen-2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>[-121.16313171386719]</td>
<td>MNGNTSWRAANTALLSVTAVEAPVVVTSDELDERLAGPMARLGIRA...</td>
<td>-121.163132</td>
<td>gen-3</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>[-133.42160034179688]</td>
<td>GNATTRFANVSLLSVASRLPSRVTPSTQIQERLGAAMTRLRLPPRL...</td>
<td>-133.421600</td>
<td>gen-4</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>[-142.5845947265625]</td>
<td>MSGNATFRLTNTAILAVTTVEAPEVITSEQLDERLRATYDRIGMTP...</td>
<td>-142.584595</td>
<td>gen-5</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>Below I have plotted: * (top) the distribution of generated sequence scores (red) against all 72 sequences in our data (grey) * (bottom) generated sequence scores (red) against a subset of highly active sequences (light blue - defined as being more active than the median enzyme).</p>
<p>We can see the scores of the generated sequences are generally right-shifted, suggesting greater fitness (defined, as we saw above, as butoxy catalysis):</p>
<div class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>f, axs <span class="op">=</span> plt.subplots(<span class="dv">2</span>,<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">12</span>,<span class="dv">6</span>), sharex<span class="op">=</span><span class="va">True</span>, sharey<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Generated sequence scores"</span>)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">0</span>].set_title(<span class="st">"Generated sequences vs all experimental sequences"</span>)</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">0</span>].hist(gen_score_result.score, bins<span class="op">=</span><span class="dv">20</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, color<span class="op">=</span><span class="st">'red'</span>, label<span class="op">=</span><span class="st">'PoET sequences'</span>)<span class="op">;</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">0</span>].hist(compound_scores.score, bins<span class="op">=</span><span class="dv">20</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, color<span class="op">=</span><span class="st">'grey'</span>, label<span class="op">=</span><span class="st">'All experimental sequences'</span>)<span class="op">;</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">0</span>].legend(bbox_to_anchor<span class="op">=</span>(<span class="fl">1.05</span>, <span class="dv">1</span>), loc<span class="op">=</span><span class="st">'upper left'</span>, borderaxespad<span class="op">=</span><span class="fl">0.</span>)</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">0</span>].set_ylabel( <span class="st">"Number of sequences"</span>)</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>median_activity <span class="op">=</span> compound_scores.mean_log_activity.median()</span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">1</span>].set_title(<span class="st">"Generated sequences vs active experimental sequences"</span>)</span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">1</span>].hist(gen_score_result.score, bins<span class="op">=</span><span class="dv">20</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, color<span class="op">=</span><span class="st">'red'</span>, label<span class="op">=</span><span class="st">'PoET sequences'</span>)<span class="op">;</span></span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">1</span>].hist(compound_scores.query(<span class="st">"mean_log_activity&gt;@median_activity"</span>).score, bins<span class="op">=</span><span class="dv">20</span>, alpha<span class="op">=</span><span class="fl">0.4</span>, color<span class="op">=</span><span class="st">'lightblue'</span>, label<span class="op">=</span><span class="st">'All active sequences'</span>)<span class="op">;</span></span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">1</span>].legend(bbox_to_anchor<span class="op">=</span>(<span class="fl">1.05</span>, <span class="dv">1</span>), loc<span class="op">=</span><span class="st">'upper left'</span>, borderaxespad<span class="op">=</span><span class="fl">0.</span>)</span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">1</span>].set_ylabel( <span class="st">"Number of sequences"</span>)</span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"model score"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-33-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>We can also examine the similarity of our generated sequences; first to each other, and then to the prompt sequences. To make the plotting easier, I have named each sequence in order of descending score (so <code>gen-1</code> is the highest scoring sequence, and <code>gen-50</code> the lowest, with <code>gen-25</code> being the median scoring sequence).</p>
<div class="cell" data-execution_count="33">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>generated_sequences <span class="op">=</span> gen_score_result.sequence.tolist()</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>gen_similarity_matrix <span class="op">=</span> generate_similarity_matrix(generated_sequences)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>gen_similarity_matrix <span class="op">=</span> pd.DataFrame(gen_similarity_matrix, index<span class="op">=</span>gen_score_result.name.tolist(), columns<span class="op">=</span>gen_score_result.name.tolist())</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>sns.heatmap(gen_similarity_matrix, cmap<span class="op">=</span><span class="st">'coolwarm'</span>, vmin<span class="op">=</span><span class="dv">0</span>, vmax<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Pairwise sequence Similarity (Generated sequences)"</span>)<span class="op">;</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-34-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="34">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Sequence pairwise-similarity statistics (generated sequences):"</span>)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss">Mean: </span><span class="sc">{</span>np<span class="sc">.</span>mean(gen_similarity_matrix.values[np.triu_indices(gen_similarity_matrix.shape[<span class="dv">0</span>], k<span class="op">=</span><span class="dv">1</span>)])<span class="sc">:.2f}</span><span class="ss"> </span><span class="ch">\n\t</span><span class="ss">Min: </span><span class="sc">{</span>np<span class="sc">.</span><span class="bu">min</span>(gen_similarity_matrix.values[np.triu_indices(gen_similarity_matrix.shape[<span class="dv">0</span>], k<span class="op">=</span><span class="dv">1</span>)])<span class="sc">:.2f}</span><span class="ss"> </span><span class="ch">\n\t</span><span class="ss">Max: </span><span class="sc">{</span>np<span class="sc">.</span><span class="bu">max</span>(gen_similarity_matrix.values[np.triu_indices(gen_similarity_matrix.shape[<span class="dv">0</span>], k<span class="op">=</span><span class="dv">1</span>)])<span class="sc">:.2f}</span><span class="ss"> "</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Sequence pairwise-similarity statistics (generated sequences):
    Mean: 0.59 
    Min: 0.50 
    Max: 0.78 </code></pre>
</div>
</div>
<p>We can see our 50 generated sequences are quite diverse, they have a mean pairwise similarity of &lt;0.6, and a range of 0.5-0.78.</p>
<div class="cell" data-execution_count="35">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>seed_gen_sim <span class="op">=</span> []</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i,seq <span class="kw">in</span> <span class="bu">enumerate</span>(generated_sequences):</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>    seed_gen_sim.append( calculate_similarity(msa_seed, seq) )</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>gen_score_result[<span class="st">'seed_similarity'</span>] <span class="op">=</span> seed_gen_sim</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Let’s see how they compare to the seed sequence we used to generate the MSA and prompt:</p>
<div class="cell" data-execution_count="36">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>f, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">5</span>))</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f"Sequence similarity: Generated vs </span><span class="sc">{</span>msa_seed_id<span class="sc">}</span><span class="ss"> seed "</span>)</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>sns.barplot(data<span class="op">=</span>gen_score_result, x<span class="op">=</span><span class="st">"name"</span>, y<span class="op">=</span><span class="st">"seed_similarity"</span>, ax<span class="op">=</span>ax, color<span class="op">=</span><span class="st">'darkblue'</span>)</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">90</span>)<span class="op">;</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Sequence name"</span>)</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Sequence similarity"</span>)<span class="op">;</span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a><span class="co"># add hline at 1 </span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>ax.axhline(<span class="dv">1</span>, ls<span class="op">=</span><span class="st">'--'</span>, color<span class="op">=</span><span class="st">'red'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-37-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Sequences are on average ~60% similar to the seed sequence. This is roughly in line with our expectations, as we saw our diverse group of thiolases that catalyze <em>all</em> 14 compounds have approximately 50% sequence similarity. We would expect <code>butoxy</code> specific enzymes to be somewhat higher than that broad thiolase baseline.</p>
<p>We can also see below that there is no significant relationship between the similarity to the seed and the model score for the generated sequences: this suggests the prompt is doing a good job (and not overfitting):</p>
<div class="cell" data-execution_count="37">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>r,p <span class="op">=</span> spearmanr(gen_score_result[<span class="st">'score'</span>], gen_score_result[<span class="st">'seed_similarity'</span>])</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Spearman correlation between score and seed similarity: </span><span class="sc">{</span>r<span class="sc">:.2f}</span><span class="ss">, p=</span><span class="sc">{</span>p<span class="sc">:.2f}</span><span class="ss">"</span>)</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>f, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">12</span>,<span class="dv">6</span>), sharex<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> axs</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a><span class="co"># full results</span></span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>sns.regplot(data<span class="op">=</span>gen_score_result,</span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>            x<span class="op">=</span><span class="st">'seed_similarity'</span>,</span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>            y<span class="op">=</span><span class="st">'score'</span>,</span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a>            scatter_kws<span class="op">=</span>{<span class="st">'s'</span>: <span class="dv">5</span>, <span class="st">'alpha'</span>: <span class="fl">0.5</span>},</span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a>            n_boot<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb54-12"><a href="#cb54-12" aria-hidden="true" tabindex="-1"></a>            robust<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb54-13"><a href="#cb54-13" aria-hidden="true" tabindex="-1"></a>            ax<span class="op">=</span>ax )</span>
<span id="cb54-14"><a href="#cb54-14" aria-hidden="true" tabindex="-1"></a>ax .set_xlabel(<span class="st">"Similarity to seed"</span>)</span>
<span id="cb54-15"><a href="#cb54-15" aria-hidden="true" tabindex="-1"></a>ax .set_ylabel(<span class="st">"Model score"</span>)</span>
<span id="cb54-16"><a href="#cb54-16" aria-hidden="true" tabindex="-1"></a><span class="co"># add spearmanr r and p results to graph</span></span>
<span id="cb54-17"><a href="#cb54-17" aria-hidden="true" tabindex="-1"></a>ax .text(<span class="fl">0.05</span>, <span class="fl">0.9</span>, <span class="ss">f"spearman r=</span><span class="sc">{</span>r<span class="sc">:.2f}</span><span class="ch">\n</span><span class="ss">p-value =</span><span class="sc">{</span>p<span class="sc">:.2f}</span><span class="ss">"</span>, transform<span class="op">=</span>plt.gca().transAxes, size<span class="op">=</span><span class="dv">14</span>)<span class="op">;</span></span>
<span id="cb54-18"><a href="#cb54-18" aria-hidden="true" tabindex="-1"></a>ax.set_title(<span class="ss">f"Seed sequence similarity vs model score"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Spearman correlation between score and seed similarity: 0.10, p=0.49</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-38-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="38">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> multiprocessing <span class="im">import</span> Pool</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> calculate_similarity_wrapper(args):</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> calculate_similarity(<span class="op">*</span>args)</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_similarity_distinct(sequences1, sequences2):</span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Generate a similarity matrix for a list of sequences"""</span></span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>    num_sequences1 <span class="op">=</span> <span class="bu">len</span>(sequences1)</span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a>    num_sequences2 <span class="op">=</span> <span class="bu">len</span>(sequences2)</span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> Pool() <span class="im">as</span> p:</span>
<span id="cb56-12"><a href="#cb56-12" aria-hidden="true" tabindex="-1"></a>        similarity_matrix <span class="op">=</span> np.array(p.<span class="bu">map</span>(calculate_similarity_wrapper, [(seq1, seq2) <span class="cf">for</span> seq1 <span class="kw">in</span> sequences1 <span class="cf">for</span> seq2 <span class="kw">in</span> sequences2]))</span>
<span id="cb56-13"><a href="#cb56-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb56-14"><a href="#cb56-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> similarity_matrix.reshape(num_sequences1, num_sequences2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Lastly, let’s look at the sequence similarity between the prompt sequences (below heatmap, Y axis) and the generated sequences (X axis):</p>
<div class="cell" data-execution_count="39">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>compared_sim <span class="op">=</span> generate_similarity_distinct(generated_sequences , promptdf.sequence.tolist() )</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>compared_sim <span class="op">=</span> pd.DataFrame(compared_sim, index<span class="op">=</span>gen_score_result.name.tolist(), columns<span class="op">=</span>promptdf.name.tolist()).T</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>f, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">8</span>,<span class="dv">12</span>), sharex<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>sns.heatmap(compared_sim, cmap<span class="op">=</span><span class="st">'coolwarm'</span>, vmin<span class="op">=</span><span class="dv">0</span>, vmax<span class="op">=</span><span class="dv">1</span>, ax<span class="op">=</span>ax)</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Pairwise sequence Similarity"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-40-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="40">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Summary statistics for pairwise generated v prompt comparison: "</span>)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss"> Mean similarity: </span><span class="sc">{</span>compared_sim<span class="sc">.</span>mean(axis<span class="op">=</span><span class="va">None</span>)<span class="sc">:.2f}</span><span class="ss">"</span>)</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss"> Max similarity: </span><span class="sc">{</span>compared_sim<span class="sc">.</span><span class="bu">max</span>(axis<span class="op">=</span><span class="va">None</span>)<span class="sc">:.2f}</span><span class="ss">"</span>)</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss"> Min similarity: </span><span class="sc">{</span>compared_sim<span class="sc">.</span><span class="bu">min</span>(axis<span class="op">=</span><span class="va">None</span>)<span class="sc">:.2f}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Summary statistics for pairwise generated v prompt comparison: 
     Mean similarity: 0.58
     Max similarity: 0.94
     Min similarity: 0.32</code></pre>
</div>
</div>
<p>Again, we can see mean similarity is low (within expectations).</p>
<p>There are a few hotspots for the top sequences, I’ve investigated these below by plotting the maximum similarity to any prompt sequence, for each generated sequence:</p>
<div class="cell" data-execution_count="41">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>gen_score_result <span class="op">=</span> gen_score_result.merge( compared_sim.<span class="bu">max</span>(axis<span class="op">=</span><span class="dv">0</span>).reset_index().rename(columns<span class="op">=</span>{<span class="st">'index'</span>:<span class="st">'name'</span>, <span class="dv">0</span>:<span class="st">'max_prompt_similarity'</span>}) , on<span class="op">=</span><span class="st">'name'</span>, how<span class="op">=</span><span class="st">'left'</span>)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> compared_sim.<span class="bu">max</span>(axis<span class="op">=</span><span class="dv">0</span>).plot(kind<span class="op">=</span><span class="st">'bar'</span>, figsize<span class="op">=</span>(<span class="dv">15</span>, <span class="dv">5</span>))<span class="op">;</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Max. similarity"</span>)</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>ax.axhline(<span class="dv">1</span>, ls<span class="op">=</span><span class="st">'--'</span>, color<span class="op">=</span><span class="st">'red'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)<span class="op">;</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Generated sequence similarity to any prompt sequence"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-42-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="42">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Average max similarity to prompt across all generated sequences: </span><span class="sc">{</span>compared_sim<span class="sc">.</span><span class="bu">max</span>(axis<span class="op">=</span><span class="dv">0</span>)<span class="sc">.</span>mean()<span class="sc">:.2f}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Average max similarity to prompt across all generated sequences: 0.75</code></pre>
</div>
</div>
<p>We see there appears to be a general trend of higher scoring sequences being more closely related to a prompt sequence (above). This is to be expected, our scoring is defined by our prompt, so sequences more similar to the prompt will score higher (by definition). We could tweak additional parameters in the PoET generate workflow if we wanted to increase, or decrease the “creativity” of the model (for example temperature, or top p).</p>
<p>Let’s take two sequences; * the best scoring (<code>gen-1</code>) * the median scoring (<code>gen-25</code>)</p>
<p>And further investigate their relationship to the prompt.</p>
<p>First, we can see the full results for prompt sequence similarity for the best sequence (<code>gen-1</code>) below. Note that for this sequence, the mean similarity to the prompt sequences is 56% (range: 46-94%).</p>
<div class="cell" data-execution_count="43">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>best_generated_sequence <span class="op">=</span> gen_score_result.sequence.tolist()[<span class="dv">0</span>]</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Prompt similarity summary for gen-1:"</span>)</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss">mean: </span><span class="sc">{</span>compared_sim[<span class="st">'gen-1'</span>]<span class="sc">.</span>mean()<span class="sc">:.2f}</span><span class="ss">"</span>)</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss">min: </span><span class="sc">{</span>compared_sim[<span class="st">'gen-1'</span>]<span class="sc">.</span><span class="bu">min</span>()<span class="sc">:.2f}</span><span class="ss">"</span>)</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss">max: </span><span class="sc">{</span>compared_sim[<span class="st">'gen-1'</span>]<span class="sc">.</span><span class="bu">max</span>()<span class="sc">:.2f}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Prompt similarity summary for gen-1:
    mean: 0.56
    min: 0.46
    max: 0.94</code></pre>
</div>
</div>
<div class="cell" data-execution_count="44">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>f, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">5</span>,<span class="dv">15</span>), sharex<span class="op">=</span><span class="va">True</span>, sharey<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>sns.barplot(data<span class="op">=</span>compared_sim[<span class="st">'gen-1'</span>].sort_values(ascending<span class="op">=</span><span class="va">False</span>).reset_index(), x<span class="op">=</span><span class="st">'gen-1'</span>, y<span class="op">=</span><span class="st">'index'</span>, ax<span class="op">=</span>axs, color<span class="op">=</span><span class="st">'darkblue'</span>, orient<span class="op">=</span><span class="st">'h'</span>)</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>axs.set_ylabel(<span class="st">""</span>)</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>plt.yticks(size<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>axs.set_xlabel(<span class="st">"Similarity"</span>)</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>axs.set_title(<span class="st">"Best generated sequence vs prompt"</span>)<span class="op">;</span></span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>axs.axvline(<span class="dv">1</span>, ls<span class="op">=</span><span class="st">'--'</span>, color<span class="op">=</span><span class="st">'red'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-45-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="45">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Best-generated sequence, closest prompt sequence:"</span>)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>promptdf[promptdf.accnumber<span class="op">==</span>compared_sim[<span class="st">'gen-1'</span>].idxmax()]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Best-generated sequence, closest prompt sequence:</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="45">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">accnumber</th>
<th data-quarto-table-cell-role="th">organism</th>
<th data-quarto-table-cell-role="th">lineage</th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">16</td>
<td>UPI00059046A7</td>
<td>Mobilicoccus massiliensis</td>
<td>[]</td>
<td>UPI00059046A7</td>
<td>MSENASYRLANCSILSVSTVEAPEVVTSDHLDELLQPAYDTTGMMA...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>The best scoring sequence actually shares 93.5% identity with the <em>M. massiliensis</em> we saw previously (and know to be an active <code>butoxy</code> thiolase). We therefore expect this sequence to perform well in any wetlab experiment.</p>
<p>Turning to the median scoring sequence:</p>
<div class="cell" data-execution_count="46">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>f, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">5</span>,<span class="dv">15</span>), sharex<span class="op">=</span><span class="va">True</span>, sharey<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>median_name <span class="op">=</span> <span class="ss">f'gen-</span><span class="sc">{</span><span class="bu">len</span>(gen_score_result)<span class="op">//</span><span class="dv">2</span><span class="sc">}</span><span class="ss">'</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>sns.barplot(data<span class="op">=</span>compared_sim[median_name].sort_values(ascending<span class="op">=</span><span class="va">False</span>).reset_index(), x<span class="op">=</span>median_name, y<span class="op">=</span><span class="st">'index'</span>, ax<span class="op">=</span>axs, color<span class="op">=</span><span class="st">'darkblue'</span>, orient<span class="op">=</span><span class="st">'h'</span>)</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>axs.set_ylabel(<span class="st">""</span>)</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>axs.set_xlabel(<span class="st">"Similarity"</span>)</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>axs.set_title(<span class="st">"Median generated sequence vs prompt"</span>)</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>axs.axvline(<span class="dv">1</span>, ls<span class="op">=</span><span class="st">'--'</span>, color<span class="op">=</span><span class="st">'red'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)<span class="op">;</span></span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()<span class="op">;</span></span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-47-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="47">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>median_generated_sequence <span class="op">=</span> gen_score_result.sequence.tolist()[<span class="bu">len</span>(gen_score_result)<span class="op">//</span><span class="dv">2</span>]</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Prompt similarity summary for gen-25:"</span>)</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss">mean: </span><span class="sc">{</span>compared_sim[<span class="st">'gen-25'</span>]<span class="sc">.</span>mean()<span class="sc">:.2f}</span><span class="ss">"</span>)</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss">min: </span><span class="sc">{</span>compared_sim[<span class="st">'gen-25'</span>]<span class="sc">.</span><span class="bu">min</span>()<span class="sc">:.2f}</span><span class="ss">"</span>)</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss">max: </span><span class="sc">{</span>compared_sim[<span class="st">'gen-25'</span>]<span class="sc">.</span><span class="bu">max</span>()<span class="sc">:.2f}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Prompt similarity summary for gen-25:
    mean: 0.60
    min: 0.49
    max: 0.71</code></pre>
</div>
</div>
<div class="cell" data-execution_count="48">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Median-generated sequence, closest prompt sequence:"</span>)</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>promptdf[promptdf.accnumber<span class="op">==</span>compared_sim[<span class="st">'gen-25'</span>].idxmax()]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Median-generated sequence, closest prompt sequence:</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="48">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">accnumber</th>
<th data-quarto-table-cell-role="th">organism</th>
<th data-quarto-table-cell-role="th">lineage</th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">19</td>
<td>A0A1H9FJA5</td>
<td>Microlunatus flavus</td>
<td>[Bacteria, Actinomycetota, Propionibacteriales...</td>
<td>A0A1H9FJA5</td>
<td>MTGNATYELKDTAVLSVCAVEAPVVVTSEAFDERLMRSYERTGLKP...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>The median scoring sequence shows a more uniform similarity to prompt sequences (mean: 60%, range: 49-71%). Note that the closest prompt sequence derives from <em>Microlunatus flavus</em>, but at 71% similarity we expect significant divergence from this organism’s thiolase.</p>
<div class="cell" data-execution_count="49">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"M. flavus in Robinson dataset: </span><span class="sc">{</span>promptdf[promptdf.accnumber<span class="op">==</span>compared_sim[<span class="st">'gen-25'</span>].idxmax()]<span class="sc">.</span>sequence<span class="sc">.</span>tolist()[<span class="dv">0</span>] <span class="kw">in</span> data<span class="sc">.</span>sequence<span class="sc">.</span>tolist()<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>M. flavus in Robinson dataset: False</code></pre>
</div>
</div>
<p>The sequences selected above represent two good case studies for futher evaluation. One sequence is the most fit sequence generated by PoET, whilst the median sequence should more closely reflect the bulk of the data. We will investigate these structurally below.</p>
</section>
<section id="investigating-generated-thiolase-structures" class="level3">
<h3 class="anchored" data-anchor-id="investigating-generated-thiolase-structures">Investigating generated thiolase structures</h3>
<p>Robinson reports a conserved Cys across all thiolases that is buried in the active site and vital for activity, we can see this below, at position 129 in the MSA seed <em>Kytococcus sedentarius</em> structure from <a href="https://www.uniprot.org/uniprotkb/C7NEV1/entry#structure">uniprot</a> (note this is an alphafold prediction of the enzyme, as the structure is yet to be solved experimentally).</p>
<div class="cell" data-execution_count="50">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>[<span class="ss">f"Cys</span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">"</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(msa_seed)) <span class="cf">if</span> msa_seed[i]<span class="op">==</span><span class="st">'C'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="50">
<pre><code>['Cys129', 'Cys225', 'Cys280', 'Cys334', 'Cys335']</code></pre>
</div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/2aac5a78-1-AF-C7NEV1-F1-MODEL_V4.PDB.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">AF-C7NEV1-F1-MODEL_V4.PDB.png</figcaption>
</figure>
</div>
<p>Note the key Cys129-His269 interaction in the active site below (vital for the thiolase “ping-pong” reaction mechanism). As well as the stabilising non-covalent interaction between Gly301 and the same His269. It is likely that this stabilising His-Gly hydrogen bonding pattern has a role in active site dynamics.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/c014dbe1-1-image.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">image.png</figcaption>
</figure>
</div>
<p>First, since we know the active site Cys is so important to catalytic activity, and that loss of this Cys would completely ablate activity, let’s examine the Cys residues in our two generated sequences of interest:</p>
<div class="cell" data-execution_count="51">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>seq_of_interest <span class="op">=</span> [<span class="st">'gen-1'</span>, <span class="st">'gen-25'</span>] <span class="co"># best scored 3 + median</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> name <span class="kw">in</span> seq_of_interest:</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>    seq <span class="op">=</span> gen_score_result.loc[gen_score_result.name<span class="op">==</span>name, <span class="st">'sequence'</span>].tolist()[<span class="dv">0</span>]</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> calculate_similarity(msa_seed, seq)</span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Sequence similarity of </span><span class="sc">{</span>msa_seed_id<span class="sc">}</span><span class="ss"> and </span><span class="sc">{</span>name<span class="sc">}</span><span class="ss">= </span><span class="sc">{</span>x<span class="sc">:.2f}</span><span class="ss">"</span>)</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Cys positions for </span><span class="sc">{</span>name<span class="sc">}</span><span class="ss">:"</span>, [i<span class="op">+</span><span class="dv">1</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(seq)) <span class="cf">if</span> seq[i]<span class="op">==</span><span class="st">'C'</span>])</span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Sequence similarity of Kytococcus sedentarius and gen-1= 0.56
Cys positions for gen-1: [120, 129, 224, 277, 279]

Sequence similarity of Kytococcus sedentarius and gen-25= 0.66
Cys positions for gen-25: [98, 120, 129, 225, 335]
</code></pre>
</div>
</div>
<p>We can see above that both our best, and median sequence, have a conserved Cys129.</p>
</section>
<section id="structural-analysis" class="level3">
<h3 class="anchored" data-anchor-id="structural-analysis">Structural analysis</h3>
<section id="best-scoring-generated-sequence" class="level4">
<h4 class="anchored" data-anchor-id="best-scoring-generated-sequence">Best scoring generated sequence</h4>
<p>For a more thorough evaluation we can also fold these sequences with <a href="https://www.deepmind.com/research/highlighted-research/alphafold">Alphafold</a>, and run a <a href="http://cospi.iiserpune.ac.in/click">pairwise structural aligner</a> with the original seed sequence.</p>
<p>We can see the results below for the best generated sequence, with the active site Cys highlighted in the enzyme’s core: * <font color="green">seed (green)</font> * <font color="darkorange">generated (orange)</font>:</p>
<p>We can see there is an overwhelming structural similarity between our seed and our best generated sequence, despite only having 56% sequence identity.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/b4e533c0-1-Robinson_best_gen_aligned.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Robinson_best_gen_aligned.png</figcaption>
</figure>
</div>
<p>Zooming in on the active site, we see the same Cys-His interaction as the seed sequence’s structure (numbered Cys129-His268 in this generated sequence).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/2864395e-1-STR2_AF_C7NEV1_F1_MODEL_V4-STR1_ROBINSON_F448D_0_UNRELAXED_RANK_001_ALPHAFOLD2_PTM_MODEL_5_SEED_000.1.PDB-STR1_ROBINSON_F448D_0_UNRELAXED_RANK_001_ALPHAFOLD2_PTM_MODEL_5_SEED_000-STR2_AF_C7NEV1_F1_MODEL_V4.1.PDB.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">STR2_AF_C7NEV1_F1_MODEL_V4-STR1_ROBINSON_F448D_0_UNRELAXED_RANK_001_ALPHAFOLD2_PTM_MODEL_5_SEED_000.1.PDB-STR1_ROBINSON_F448D_0_UNRELAXED_RANK_001_ALPHAFOLD2_PTM_MODEL_5_SEED_000-STR2_AF_C7NEV1_F1_MODEL_V4.1.PDB.png</figcaption>
</figure>
</div>
<p>Looking at the full ensemble of 5 models that AlphaFold produced (below), we can see there is overall agreement for both the Cys (yellow) and Histidine (blue) in the generated sequence across predictions, this may suggest there is little uncertainty or flexibility around these residues.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/d91b042a-1-image.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">image.png</figcaption>
</figure>
</div>
<p>This matches both the <em>K. sedentarius</em> and <em>M. massiliensis</em> AphaFold predictions. I’ve reproduced the <em>M. massiliensis</em> 5 model ensemble below for illustration:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/ac8e17b8-1-image.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">image.png</figcaption>
</figure>
</div>
<p>An interesting difference however, is that the His269-Gly301 stabilizing non-covalent interaction in the seed structure has been replaced by a His268-Gln269 interaction in our generated structure:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/a925e525-1-Robinson_best_generated_zoom2.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Robinson_best_generated_zoom2.png</figcaption>
</figure>
</div>
<p>Overall, we can see a high degree of structural similarity, with a few key active site differences. The overall RMSD between the seed and best generated sequence is &lt;1A:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/884e4674-1-image.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">image.png</figcaption>
</figure>
</div>
<p>We can also compare this generated sequence to it’s closest prompt sequence (which had a 94% sequence identity):</p>
<p>First at the sequence level (below) which shows significant similarity:</p>
<div class="cell" data-execution_count="52">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>most_similar_prompt_sequence <span class="op">=</span> promptdf.loc[promptdf.name <span class="op">==</span> compared_sim[<span class="st">'gen-1'</span>].sort_values().index.tolist()[<span class="op">-</span><span class="dv">1</span>] , <span class="st">'sequence'</span>].tolist()[<span class="dv">0</span>]</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>aligned <span class="op">=</span> align_report(most_similar_prompt_sequence, best_generated_sequence)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>MSENAS-YRLANC-SI-LSVSTV-EAPEVVTSDHLDELLQPAYDT-TGMMAGQIERLAGVRERRWFPKDADYTR-GAVEAGRQAI-ADARIDADRIGLIVNASVTRPHLEPGI-SAMV-HAELELPRSCIAFDVTNACLGVVNSLQIAG-TMIDAGQIEY--ALIVAS-EGARQMQEASIARLV-DGGSRQDVKDAFATMTLGSGAVGIVLGRS-DRHPEGHQ-IIGGITRAGT-EHHELCIGGMDGMRTDAS-KLFVEGLSLATDAWAEAKGEWDWEEMDWYVAHQTSTAHIACLCDTLKLP-A-EKFPVTVDTYGNIGPVALPFTLGMYQH-KMRKGDRVLLMGIGSGLNTSMT-EVRW
|||||  |||||  |  |||||  ||||||||||||||||||||  |||||||||||||||||||||||||||  |||||||||  ||||||||||||||||||||||||||  || | |||||||||||||||||||||||||||||  ||||||||    |||| | ||||||||||||||  ||||||||||||||||||||||||||||  |||||||  |||||||||  |||||||||||||||||  |||||||||||||||||||||||||||||||||||||||||||||||   | ||||||||||||||||||||||||||  |||||||||||||||||||||  ||||
MSENA-TYRLAN-TS-VLSVST-IEAPEVVTSDHLDELLQPAYD-STGMMAGQIERLAGVRERRWFPKDADYT-QGAVEAGRQA-MADARIDADRIGLIVNASVTRPHLEPG-VSA-VVHAELELPRSCIAFDVTNACLGVVNSLQIA-ATMIDAGQI--DNALIV-SSEGARQMQEASIARL-LDGGSRQDVKDAFATMTLGSGAVGIVLGR-ADRHPEGH-RIIGGITRAG-SEHHELCIGGMDGMRTDA-RKLFVEGLSLATDAWAEAKGEWDWEEMDWYVAHQTSTAHIACLCDTLK--VAPEKFPVTVDTYGNIGPVALPFTLGMYQ-QKMRKGDRVLLMGIGSGLNTSM-AEVRW
  Score=317
</code></pre>
</div>
</div>
<div class="cell" data-execution_count="53">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>start <span class="op">=</span> <span class="dv">250</span> </span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>stop <span class="op">=</span> <span class="dv">310</span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Alignment at ~250-310 region:"</span>)</span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">3</span>):</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(aligned.split(<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)[i][start:stop])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Alignment at ~250-310 region:
DAS-KLFVEGLSLATDAWAEAKGEWDWEEMDWYVAHQTSTAHIACLCDTLKLP-A-EKFP
||  |||||||||||||||||||||||||||||||||||||||||||||||   | ||||
DA-RKLFVEGLSLATDAWAEAKGEWDWEEMDWYVAHQTSTAHIACLCDTLK--VAPEKFP</code></pre>
</div>
</div>
<p>Interestingly, the <em>M. massiliensis</em> structure (below, AlphaFold), shows a different hydrogen bonding pattern for His268, despite the sequence conservation (above). The <em>M. massiliensis</em> structure lacks any stabilising hydrogen bonding between His268 and adjacent residues. This differs to both the <em>K. sedentarius</em> structure (where Gly301 provides support) and the generated sequence above (where Gln269 provides support). Given that the <em>K. sedentarius</em> and <em>M. massiliensis</em> thiolases were amongst the most active for <code>butoxy</code>, there is reason to believe this generated sequence, which borrows from both, would also be highly active against <code>butoxy</code>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/c9e74650-1-STR1_MOBILICOCCUSMASSILIENSIS_17287_RELAXED_RANK_001_ALPHAFOLD2_PTM_MODEL_5_SEED_000-STR2_ROBINSON_F448D_0_UNRELAXED_RANK_001_ALPHAFOLD2_PTM_MODEL_5_SEED_000.1.PDB-STR2_ROBINSON_F448D_0_UNRELAXED_RANK_001_ALPHAFOLD.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">STR1_MOBILICOCCUSMASSILIENSIS_17287_RELAXED_RANK_001_ALPHAFOLD2_PTM_MODEL_5_SEED_000-STR2_ROBINSON_F448D_0_UNRELAXED_RANK_001_ALPHAFOLD2_PTM_MODEL_5_SEED_000.1.PDB-STR2_ROBINSON_F448D_0_UNRELAXED_RANK_001_ALPHAFOLD.png</figcaption>
</figure>
</div>
<p>Overall, we can see there is a strong structural similarity between our best scoring generated sequence and both the seed sequence and it’s closest prompt sequence; this is reflected in the alignment results, which show a &lt;1A RMSD. Additionally we see strong evidence for conservation at the active site, specifically the Cys-His interaction. Interestingly the stabilising hydrogen bonding pattern differs between all three sequences. And despite the strong sequence similarity between our sequence and <em>M. massiliensis</em>, our structure has elements more reminscent of <em>K. sedentarius</em> (e.g.&nbsp;the stabilising hydrogen bonding around the active site Histidine). Given these shared elements from two of the best <code>butoxy</code> thiolases in our dataset, we would expect this sequence to perform well in any wetlab assay.</p>
</section>
<section id="median-scoring-sequence" class="level4">
<h4 class="anchored" data-anchor-id="median-scoring-sequence">Median scoring sequence</h4>
<p>Let’s now look at the median generated sequence, which has a much lower maximum sequence similarity to any sequence in the prompt (71%). We will follow the same workflow, and look for similar patterns in the active site that might indicate specificity or activity:</p>
<p>First, we can see the overall alignment between this sequence and the seed sequence is good, RMSD is again &lt;1A.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/4e6e0e9f-1-image.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">image.png</figcaption>
</figure>
</div>
<p>Focussing on the active site (below), we see His269 and Cys129 are conserved again, likely preserving the core “ping-pong” catalytic mechanism. However, the Gln269-His268 interaction that we saw in the best generated sequence, is absent here (reminiscent of <em>M. massiliensis</em>).</p>
<p>There is a Gly301 in the vicinity of the active site Histidine, but this Glycine is distanced from His269, suggesting those stabilizing hydogen bonds we saw in the seed structure will be transient or non-existent here. This may have repercussions for active site volume and solvent accessibility, and therefore activity and specificity (as noted by Robinsin et al.&nbsp;2020).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/1d165ed8-1-image.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">image.png</figcaption>
</figure>
</div>
<p>Examining the AlphaFold ensemble of 5 models (below), we see clear agreement on the Cys129 orientation, but less agreement on the positioning of the active site Histidine (blue) than either <em>K. sedentarius</em>, <em>M. massiliensis</em> or the best generated sequence. This may suggest more flexibility around His269, potentially due to the weaker non-covalent interactions discussed above:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/daf2c0cf-1-image.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">image.png</figcaption>
</figure>
</div>
<p>If we examine the sequence similarity between this median generated sequence and it’s most similar prompt sequence (below) we see less conservation overall, as expected from its low sequence similarity percentage:</p>
<div class="cell" data-execution_count="54">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>most_similar_prompt_sequence <span class="op">=</span> promptdf.loc[promptdf.name <span class="op">==</span> compared_sim[<span class="st">'gen-25'</span>].sort_values().index.tolist()[<span class="op">-</span><span class="dv">1</span>] , <span class="st">'sequence'</span>].tolist()[<span class="dv">0</span>]</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>aligned <span class="op">=</span> align_report(most_similar_prompt_sequence, median_generated_sequence)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>MTG-NAT----YELKDT---AVL-SVCA-VEA--PV-VVTSEAFDERLMRSYE-RT-G--LK----PGMVEQ-L----AGVRE-RRWWP--ED--VTFADA--A-AM--AGAK--ALAEAGV---DPSQVGLLI-NTSVC-RDGIE---PAN--A-AGVHDQ--LGLAT-SCL---NFDVT-NACLGFL-NG-IQ-LAGS--LIDS-G-QVE-YALLV----DAEGL-RGL-QESTLA---RLAQ--DDA-TA-EDLKLQF-A---TLTLGSGSAAM--VL-GR-ASE--HPEG--HR-VTGGVS-RSG--TQHKD--LCWA--DLTQ---MR-TDSRA--LFEA--GIE-L-----AGQTWQ-DALEGREWADA--DWFIAHQTSV-----V-HLQAMAKV--VG-----VPDEKFPKT-V-------P-TF----GNL---ASAAV-PF-TL---AKQVPMIK---PGQQV--ILL--GI-GSGLNTSFA--EII--W
||  |      |        | | ||   | |  |  |||| |  | |    | |  |  ||    |    | |    ||||  ||     |   |   |   | |   ||    ||||||    |  ||| || ||||  |   |   |    | | ||    |||   |     ||||  ||||||  |  |  ||    ||   | ||  || ||    ||    |   ||   |   ||    ||  |  |    || |   ||||||| ||   |  |  |    ||    || | |||  |    |||    ||    |      |  ||  |  |     |   |     |   |  ||   |    |  ||      |      | | |    |  |      |      |  |       | |     ||    ||    |  ||   |      |   ||     | |  |  ||||||  |  |    |
MT-RN--SNINY-----RNSA-LLSV--SV-ADAP-EVVTS-A--E-L----EQR-LGPTLKRLRLP----QGLLSRVAGVR-ARR---SYE-PGV---D-FEAGA-VEAG--RRALAEAG-IRTD--QVG-LIINTSV-TR---EHLEP--SVAVA-VH--HRLGL--PS--SAMNFDV-ANACLGF-VN-AI-TLA--ATLI--EGGQV-DYA-LVVNGEDA---QR--VQE---ATIERL--NTDD-VT-RE----QFMAEFATLTLGSG-AA-AAV-VG-PA--DRHP--DSHRIV-GGV-TR--AATQH--HALC--IGD---HHGM-FTD--AQGL---LRG--GLDLVLDA---W-SDA---R----ANWDW------S-QMDRYVTH-Q----VSSV-HTNSIV------K-AVNIDADRVPVT-YPELGN-VGPAS---LP-LTLSREA------KHLAPG---DRI-LCMG-VGSGLNT--AMLE--LEW
  Score=180
</code></pre>
</div>
</div>
<p>Interestingly, we see evidence for strong conservation around the Cys129 residue (see below: the NACL motif is conserved):</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/96b2e75a-1-image.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">image.png</figcaption>
</figure>
</div>
<p>But less conservation around the His269 at the sequence level:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/d6ff65b9-1-image.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">image.png</figcaption>
</figure>
</div>
<p>This pattern of conservation (structure and sequence) at the Cys129 position, with non-conservation at the His269 position is likely to have implications for activity.</p>
<p>Overall, the generate functionality of PoET has yielded a sequence with low identity but high structural fidelity. This median scoring sequence shows only 66% identity to the seed, and 71% to the closest prompt sequence. Despite this low sequence level similarity, the structures are remarkably similar, and the active sites have been preserved very carefully. Furthermore, the changes in secondary support for the active site are likely to effect protein dynamics and activity and warrant further study.</p>
</section>
</section>
<section id="generated-sequences-summary" class="level3">
<h3 class="anchored" data-anchor-id="generated-sequences-summary">Generated sequences summary <a class="anchor" id="3_3"></a></h3>
<p>We have generated 50 sequences <em>de novo</em> by using an MSA of our <code>butoxy</code> catalysing enzyme as a prompt. We previously saw that sequence fitness scored by protsprot correlated well with <code>butoxy</code> activity (but not with <code>dimethyl</code>, or other compounds). We therefore might expect these novel sequences to show similar activity and specificity against <code>butoxy</code>.</p>
<p>Utilizing AlphaFold, we were able to show that 2 representative sequences (the best, and the median) align with the seed sequence structure to within 1A RMSD, despite having low sequence identity to the seed (50-60%) and, in the case of the median sequence, low identity to <em>any</em> prompt sequence (mean: 60%, max: 71%). Furthermore, these generated sequences preserve a vital Cys-His interaction in the active site, but differ in how they stabilize the active site His with surrounding non-covalent interactions. We hypothesize that these secondary differences may influence substrate activity and specificity and warrant further study.</p>
<p>Overall, these sequences (or a subset of them) could be progressed to wetlab evaluation to assess for improved activity over our <em>K. sedentarius</em> sequence. This would then generate more data that could be used in OpenProtein’s suite of tools to engineer increasingly active and specific thiolases!</p>
</section>
</section>
<section id="summary-1" class="level1">
<h1>Summary <a class="anchor" id="4_1"></a></h1>
<p>OpenProtein’s PoET workflow, a state-of-the-art protein language model, enables the evaluation of protein sequences without functional measurements, leveraging multiple sequence alignment to assess protein fitness and generate novel sequences.</p>
<p>Using an external thiolase dataset from <a href="https://academic.oup.com/synbio/article/5/1/ysaa004/5847618">Robinson 2020</a>, PoET demonstrated a strong correlation between model scores and experimental butoxy catalytic activity. Furthermore, the novel sequences generated in the second half of this demo showed structural similarity to the seed sequence and potential for improved activity, thus offering a powerful tool for protein engineering without preliminary wetlab experiments.</p>
</section>
<section id="appendix" class="level1">
<h1>Appendix <a class="anchor" id="Appendix"></a></h1>
<section id="additional-figures" class="level2">
<h2 class="anchored" data-anchor-id="additional-figures">Additional figures:</h2>
<section id="full-results-for-compound-activity-vs-model-score" class="level3">
<h3 class="anchored" data-anchor-id="full-results-for-compound-activity-vs-model-score">Full results for compound activity vs model score:</h3>
<div class="cell" data-execution_count="55">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>f, axs <span class="op">=</span> plt.subplots(<span class="bu">len</span>(compound_correlations)<span class="op">//</span><span class="dv">2</span>, <span class="dv">2</span>,  figsize<span class="op">=</span>(<span class="dv">15</span>,<span class="dv">15</span>), sharex<span class="op">=</span><span class="va">True</span>, sharey<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>axs <span class="op">=</span> axs.flatten()</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Model score vs compound activity:</span><span class="ch">\n</span><span class="st"> all compounds"</span>, y<span class="op">=</span><span class="fl">1.02</span>)</span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, compound <span class="kw">in</span>  <span class="bu">enumerate</span>(compound_correlations.index.tolist()):</span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> scoring_result.merge(data.query(<span class="st">"compound==@compound"</span>), on<span class="op">=</span><span class="st">'sequence'</span>)</span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> x.query(<span class="st">"mean_log_activity&gt;0"</span>)</span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a>    r,p <span class="op">=</span> spearmanr(x[<span class="st">'score'</span>], x[<span class="st">'mean_log_activity'</span>])  </span>
<span id="cb85-8"><a href="#cb85-8" aria-hidden="true" tabindex="-1"></a>    sns.regplot(data<span class="op">=</span>x,</span>
<span id="cb85-9"><a href="#cb85-9" aria-hidden="true" tabindex="-1"></a>                x<span class="op">=</span><span class="st">'mean_log_activity'</span>,</span>
<span id="cb85-10"><a href="#cb85-10" aria-hidden="true" tabindex="-1"></a>                y<span class="op">=</span><span class="st">'score'</span>,</span>
<span id="cb85-11"><a href="#cb85-11" aria-hidden="true" tabindex="-1"></a>                scatter_kws<span class="op">=</span>{<span class="st">'s'</span>: <span class="dv">5</span>, <span class="st">'alpha'</span>: <span class="fl">0.5</span>},</span>
<span id="cb85-12"><a href="#cb85-12" aria-hidden="true" tabindex="-1"></a>                n_boot<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb85-13"><a href="#cb85-13" aria-hidden="true" tabindex="-1"></a>                robust<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb85-14"><a href="#cb85-14" aria-hidden="true" tabindex="-1"></a>                ax<span class="op">=</span>axs[i] )</span>
<span id="cb85-15"><a href="#cb85-15" aria-hidden="true" tabindex="-1"></a>    axs[i] .set_ylabel(<span class="st">""</span>)</span>
<span id="cb85-16"><a href="#cb85-16" aria-hidden="true" tabindex="-1"></a>    axs[i].set_xlabel(<span class="st">""</span>)</span>
<span id="cb85-17"><a href="#cb85-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add spearmanr r and p results to graph</span></span>
<span id="cb85-18"><a href="#cb85-18" aria-hidden="true" tabindex="-1"></a>    axs[i].set_title(<span class="ss">f"</span><span class="sc">{</span>compound<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>r<span class="op">=</span><span class="sc">:.2f}</span><span class="ss">, </span><span class="sc">{</span>p<span class="op">=</span><span class="sc">:.2f}</span><span class="ss">"</span>)<span class="op">;</span></span>
<span id="cb85-19"><a href="#cb85-19" aria-hidden="true" tabindex="-1"></a>axs[<span class="bu">len</span>(compound_correlations)<span class="op">-</span><span class="dv">2</span>].set_xlabel(<span class="st">"log10 activity"</span>)</span>
<span id="cb85-20"><a href="#cb85-20" aria-hidden="true" tabindex="-1"></a>axs[<span class="bu">len</span>(compound_correlations)<span class="op">-</span><span class="dv">1</span>].set_xlabel(<span class="st">"log10 activity"</span>)</span>
<span id="cb85-21"><a href="#cb85-21" aria-hidden="true" tabindex="-1"></a>axs[<span class="bu">len</span>(compound_correlations)<span class="op">-</span><span class="dv">2</span>].set_ylabel(<span class="st">"Model score"</span>)</span>
<span id="cb85-22"><a href="#cb85-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-23"><a href="#cb85-23" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-56-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
</section>
<section id="additional-prompts" class="level2">
<h2 class="anchored" data-anchor-id="additional-prompts">Additional prompts:</h2>
<section id="scoring-with-default-parameters-yields-a-weaker-model" class="level3">
<h3 class="anchored" data-anchor-id="scoring-with-default-parameters-yields-a-weaker-model">Scoring with default parameters yields a weaker model: <a class="anchor" id="defaultprompt"></a></h3>
<p>The default prompt settings (as used below) are: * Use first sequence as seed to generate MSA = <code>True</code> (as before) * number of prompts to ensemble = <code>1</code> * minimum similarity to seed = <code>0.0</code> * Homology level = <code>0.8</code> * always include seed sequence: <code>False</code> * maximum number of sequences = <code>0</code> * maximum number of amino acids = <code>12000</code></p>
<div class="cell" data-execution_count="56">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>results_default_prompt <span class="op">=</span> <span class="st">"./results/Robinson2020/prots2prot-query-defaultprompt.csv"</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>scoring_defaultprompt_job_id <span class="op">=</span> <span class="st">'0d1576e5-f954-453f-ba4e-d5c218701946'</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>scoring_result_defaultprompt <span class="op">=</span> pd.read_csv(results_default_prompt)</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Reading from </span><span class="sc">{</span>scoring_defaultprompt_job_id<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>scoring_result_defaultprompt.head(<span class="dv">3</span>)    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Reading from 0d1576e5-f954-453f-ba4e-d5c218701946</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="56">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">scores</th>
<th data-quarto-table-cell-role="th">sequence</th>
<th data-quarto-table-cell-role="th">score</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>[-497.7345581054688]</td>
<td>MRAAQRYDDVVITGLAHVDAPHVVTSTELEDRLAGTLGRLRITPGL...</td>
<td>-497.734558</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>[-329.5821228027344]</td>
<td>MNVGITGIAYALPAREVTTAELQRRVAPKLPPRMFEQATGITRRRV...</td>
<td>-329.582123</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>[-232.86839294433597]</td>
<td>MDGNATTTHRNVALLSVATTMATRITTSAEIDRRLAPSLKRLKLPT...</td>
<td>-232.868393</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div class="cell" data-execution_count="57">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> scoring_result_defaultprompt.merge(data.query(<span class="st">"compound==@selected_compound"</span>), on<span class="op">=</span><span class="st">'sequence'</span>)</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> x.query(<span class="st">"mean_log_activity&gt;0"</span>)</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>r,p <span class="op">=</span> spearmanr(x[<span class="st">'score'</span>], x[<span class="st">'mean_log_activity'</span>])</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Spearman correlation between score and mean_log_activity: </span><span class="sc">{</span>r<span class="sc">:.2f}</span><span class="ss">, p=</span><span class="sc">{</span>p<span class="sc">:.2f}</span><span class="ss">"</span>)</span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>f, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">12</span>,<span class="dv">6</span>), sharex<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> axs</span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a><span class="co"># full results</span></span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a>sns.regplot(data<span class="op">=</span>x,</span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a>            x<span class="op">=</span><span class="st">'mean_log_activity'</span>,</span>
<span id="cb88-11"><a href="#cb88-11" aria-hidden="true" tabindex="-1"></a>            y<span class="op">=</span><span class="st">'score'</span>,</span>
<span id="cb88-12"><a href="#cb88-12" aria-hidden="true" tabindex="-1"></a>            scatter_kws<span class="op">=</span>{<span class="st">'s'</span>: <span class="dv">5</span>, <span class="st">'alpha'</span>: <span class="fl">0.5</span>},</span>
<span id="cb88-13"><a href="#cb88-13" aria-hidden="true" tabindex="-1"></a>            n_boot<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb88-14"><a href="#cb88-14" aria-hidden="true" tabindex="-1"></a>            robust<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb88-15"><a href="#cb88-15" aria-hidden="true" tabindex="-1"></a>            ax<span class="op">=</span>ax )</span>
<span id="cb88-16"><a href="#cb88-16" aria-hidden="true" tabindex="-1"></a>ax .set_xlabel(<span class="st">"Experimental "</span>)</span>
<span id="cb88-17"><a href="#cb88-17" aria-hidden="true" tabindex="-1"></a>ax .set_ylabel(<span class="st">"Model score"</span>)</span>
<span id="cb88-18"><a href="#cb88-18" aria-hidden="true" tabindex="-1"></a><span class="co"># add spearmanr r and p results to graph</span></span>
<span id="cb88-19"><a href="#cb88-19" aria-hidden="true" tabindex="-1"></a>ax .text(<span class="fl">0.05</span>, <span class="fl">0.9</span>, <span class="ss">f"spearman r=</span><span class="sc">{</span>r<span class="sc">:.2f}</span><span class="ch">\n</span><span class="ss">p-value =</span><span class="sc">{</span>p<span class="sc">:.2f}</span><span class="ss">"</span>, transform<span class="op">=</span>plt.gca().transAxes, size<span class="op">=</span><span class="dv">14</span>)<span class="op">;</span></span>
<span id="cb88-20"><a href="#cb88-20" aria-hidden="true" tabindex="-1"></a>ax.set_title(<span class="ss">f"Correlation between </span><span class="sc">{</span>selected_compound<span class="sc">}</span><span class="ss"> and log-liklihood score</span><span class="ch">\n</span><span class="ss"> Default prompt. "</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Spearman correlation between score and mean_log_activity: 0.26, p=0.03</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-58-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="58">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>response <span class="op">=</span> session.prots2prot.get_prompt(scoring_defaultprompt_job_id, prompt_index<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>defaultpromptdf <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(response), columns<span class="op">=</span>[<span class="st">'name'</span>,<span class="st">'sequence'</span>])</span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="co">#rename seed from 101 to uniprotkb entry </span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>defaultpromptdf.loc[defaultpromptdf.name<span class="op">==</span><span class="st">'101'</span>, <span class="st">'name'</span>] <span class="op">=</span> <span class="st">'C7NEV1'</span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Number of sequences in prompt:"</span>, <span class="bu">len</span>(defaultpromptdf))</span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Sample:"</span>)</span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>defaultpromptdf.head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Number of sequences in prompt: 40
Sample:</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="58">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>A0A1F9RVZ1</td>
<td>RYDKVYLAGLGYELAPVVVDTAEIEERLKPALAALRIAEGQIESLT...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>A0A810ML27</td>
<td>PVGITAVASVLPDREVTTADLQQRVAAASNLPLPPGMFAQATGIDR...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>A0A354ZDY2</td>
<td>KVMPLFWGYTGSACVPMALAVSRGRIRSGDLVTLVGSGVGYNAGVALL</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>A0A3N1ARH3</td>
<td>VGITGVGVSLPDRTLTTADLQDRVAAASDVVLPPGLFATMTGIGQR...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>A0A367YQ38</td>
<td>MKTDATYQLNNTAICSAVAVEAPDRVTSAWIDEQLAETYQRLGMTP...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>A0A1V5NTN6</td>
<td>FSRVFLESIGYELPHNIITSTWIEERLSPLYKKLFLQPGQLEALTG...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>A0A535XU34</td>
<td>MAGREVKYATRQLTESSLTALREGMSVDDIDQFLFHQANLRIIESV...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>UPI0003B5A6F3</td>
<td>MNGNATTRHNNVALLAVASTIAPRITTSHEIDQQLAPVLKRLRLPQ...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>A0A7X8QXU4</td>
<td>RFQKACISSIGYELGPVVVTSDALEARIAPVYKSLGLPLGQLQGLT...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>A0A0K1PEU7</td>
<td>RWSRVCIEAIAYELPDERVTSSALEARLAPVYEALRLGGGQLEALT...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div class="cell" data-execution_count="59">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Number of overlapping sequences in default args prompt and query: </span><span class="sc">{</span><span class="bu">len</span>(<span class="bu">set</span>(unique_sequences) <span class="op">&amp;</span> <span class="bu">set</span>(defaultpromptdf.sequence.unique().tolist()))<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> [i <span class="cf">for</span> i <span class="kw">in</span> unique_sequences <span class="cf">if</span> i <span class="kw">in</span> defaultpromptdf.sequence.unique().tolist()]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Number of overlapping sequences in default args prompt and query: 0</code></pre>
</div>
</div>
<div class="cell" data-execution_count="60">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>prompt_orgs <span class="op">=</span> [] </span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> accnumber <span class="kw">in</span> defaultpromptdf.name.unique()[<span class="dv">0</span>:]:</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> accnumber <span class="op">==</span> <span class="st">'101'</span>:</span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>         <span class="cf">continue</span> </span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>    r <span class="op">=</span> requests.get(<span class="ss">f"https://www.uniprot.org/uniprotkb/</span><span class="sc">{</span>accnumber<span class="sc">}</span><span class="ss">.json"</span>, timeout<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>    prompt_orgs.append((accnumber, r.json().get(<span class="st">'organism'</span>, {}).get(<span class="st">'scientificName'</span>, <span class="st">'notfound'</span>), r.json().get(<span class="st">'organism'</span>, {}).get(<span class="st">'lineage'</span>, [] )))</span>
<span id="cb94-7"><a href="#cb94-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-8"><a href="#cb94-8" aria-hidden="true" tabindex="-1"></a>defaultpromptdf <span class="op">=</span> pd.DataFrame(prompt_orgs, columns<span class="op">=</span>[<span class="st">'accnumber'</span>, <span class="st">'organism'</span>, <span class="st">'lineage'</span>]).merge(defaultpromptdf, left_on<span class="op">=</span><span class="st">'accnumber'</span>, right_on<span class="op">=</span><span class="st">'name'</span>)</span>
<span id="cb94-9"><a href="#cb94-9" aria-hidden="true" tabindex="-1"></a>defaultpromptdf.organism.value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="60">
<pre><code>organism
notfound                                                5
Elusimicrobia bacterium                                 3
Desulfobacteraceae bacterium                            2
Planctomycetota bacterium                               2
Acidobacteria bacterium                                 2
Desulfobulbaceae bacterium BRH_c16a                     1
Opitutae bacterium                                      1
Candidatus Aminicenantes bacterium                      1
Deltaproteobacteria bacterium RIFOXYD12_FULL_57_12      1
Myxococcales bacterium                                  1
Oligosphaeraceae bacterium                              1
Puniceicoccaceae bacterium                              1
Candidatus Wallbacteria bacterium HGW-Wallbacteria-1    1
Roseimaritima ulvae                                     1
Streptomyces lavendulae subsp. lavendulae               1
Acidimicrobiales bacterium                              1
Actinomycetia bacterium                                 1
Elusimicrobia bacterium GWA2_69_24                      1
Gammaproteobacteria bacterium                           1
Halobacteriovoraceae bacterium                          1
Rickettsiales bacterium                                 1
Acidimicrobium sp                                       1
candidate division KSB3 bacterium                       1
Planctomycetales bacterium                              1
Vulgatibacter incomptus                                 1
Chloroflexi bacterium                                   1
bacterium ADurb.Bin374                                  1
Desertihabitans brevis                                  1
Micromonospora sp. Llam0                                1
Firmicutes bacterium                                    1
Halobacteriovorax sp                                    1
Name: count, dtype: int64</code></pre>
</div>
</div>
<p>Interestingly, our default prompt contains neither the K. sedentarius or M. massiliensis sequence. And is drawn from a much larger range of taxonomic orders. This likely is including thiolases other than <code>butoxy</code> thiolases.</p>
<div class="cell" data-execution_count="61">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>defaultpromptdf[<span class="st">'seed_similarity'</span>] <span class="op">=</span> [calculate_similarity(msa_seed, s) <span class="cf">for</span> s <span class="kw">in</span> defaultpromptdf.sequence ]</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>promptdf[<span class="st">'seed_similarity'</span>] <span class="op">=</span> [calculate_similarity(msa_seed, s) <span class="cf">for</span> s <span class="kw">in</span> promptdf.sequence ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="62">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>f, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">12</span>,<span class="dv">6</span>), sharex<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>sns.histplot(data<span class="op">=</span>defaultpromptdf, x<span class="op">=</span><span class="st">'seed_similarity'</span>, bins<span class="op">=</span><span class="dv">20</span>, label<span class="op">=</span><span class="st">'default prompt'</span>,alpha<span class="op">=</span><span class="fl">0.7</span>, ax<span class="op">=</span>ax)</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>sns.histplot(data<span class="op">=</span>promptdf, x<span class="op">=</span><span class="st">'seed_similarity'</span>, bins<span class="op">=</span><span class="dv">20</span>, label<span class="op">=</span><span class="st">'performant prompt'</span>,alpha<span class="op">=</span><span class="fl">0.7</span>, ax<span class="op">=</span>ax)</span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>plt.legend()</span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f"Prompt sequence similarity to </span><span class="sc">{</span>msa_seed_id<span class="sc">}</span><span class="ss"> seed"</span>)</span>
<span id="cb97-6"><a href="#cb97-6" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="ss">f"Similarity"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-63-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>We can see above the prompt produced with the default args contains sequences that are far more divergent than our performant prompt used in section 2. This is likely a symptom of the taxonomic diversity and likely the cause of the poor performance with the default prompt.</p>
<div class="cell" data-execution_count="63">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Full results from default prompt args job</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>f, axs <span class="op">=</span> plt.subplots(data.compound.nunique()<span class="op">//</span><span class="dv">2</span>, <span class="dv">2</span>,  figsize<span class="op">=</span>(<span class="dv">15</span>,<span class="dv">15</span>), sharex<span class="op">=</span><span class="va">True</span>, sharey<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>axs <span class="op">=</span> axs.flatten()</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Correlation between experimental activity and model score</span><span class="ch">\n</span><span class="st"> Default prompt"</span>, y<span class="op">=</span><span class="fl">1.02</span>)</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, compound <span class="kw">in</span>  <span class="bu">enumerate</span>(data.compound.unique()):</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> scoring_result_defaultprompt.merge(data.query(<span class="st">"compound==@compound"</span>), on<span class="op">=</span><span class="st">'sequence'</span>)</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> x.query(<span class="st">"mean_log_activity&gt;0"</span>)</span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>    r,p <span class="op">=</span> spearmanr(x[<span class="st">'score'</span>], x[<span class="st">'mean_log_activity'</span>])  </span>
<span id="cb98-9"><a href="#cb98-9" aria-hidden="true" tabindex="-1"></a>    sns.regplot(data<span class="op">=</span>x,</span>
<span id="cb98-10"><a href="#cb98-10" aria-hidden="true" tabindex="-1"></a>                x<span class="op">=</span><span class="st">'mean_log_activity'</span>,</span>
<span id="cb98-11"><a href="#cb98-11" aria-hidden="true" tabindex="-1"></a>                y<span class="op">=</span><span class="st">'score'</span>,</span>
<span id="cb98-12"><a href="#cb98-12" aria-hidden="true" tabindex="-1"></a>                scatter_kws<span class="op">=</span>{<span class="st">'s'</span>: <span class="dv">5</span>, <span class="st">'alpha'</span>: <span class="fl">0.5</span>},</span>
<span id="cb98-13"><a href="#cb98-13" aria-hidden="true" tabindex="-1"></a>                n_boot<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb98-14"><a href="#cb98-14" aria-hidden="true" tabindex="-1"></a>                robust<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb98-15"><a href="#cb98-15" aria-hidden="true" tabindex="-1"></a>                ax<span class="op">=</span>axs[i] )</span>
<span id="cb98-16"><a href="#cb98-16" aria-hidden="true" tabindex="-1"></a>    axs[i] .set_ylabel(<span class="st">""</span>)</span>
<span id="cb98-17"><a href="#cb98-17" aria-hidden="true" tabindex="-1"></a>    axs[i].set_xlabel(<span class="st">""</span>)</span>
<span id="cb98-18"><a href="#cb98-18" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add spearmanr r and p results to graph</span></span>
<span id="cb98-19"><a href="#cb98-19" aria-hidden="true" tabindex="-1"></a>    axs[i].set_title(<span class="ss">f"Default prompt: </span><span class="sc">{</span>compound<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>r<span class="op">=</span><span class="sc">:.2f}</span><span class="ss">, </span><span class="sc">{</span>p<span class="op">=</span><span class="sc">:.2f}</span><span class="ss">"</span>)<span class="op">;</span></span>
<span id="cb98-20"><a href="#cb98-20" aria-hidden="true" tabindex="-1"></a>axs[<span class="bu">len</span>(compound_correlations)<span class="op">-</span><span class="dv">2</span>].set_xlabel(<span class="st">"log10 activity"</span>)</span>
<span id="cb98-21"><a href="#cb98-21" aria-hidden="true" tabindex="-1"></a>axs[<span class="bu">len</span>(compound_correlations)<span class="op">-</span><span class="dv">1</span>].set_xlabel(<span class="st">"log10 activity"</span>)</span>
<span id="cb98-22"><a href="#cb98-22" aria-hidden="true" tabindex="-1"></a>axs[<span class="bu">len</span>(compound_correlations)<span class="op">-</span><span class="dv">2</span>].set_ylabel(<span class="st">"Model score (default prompt)"</span>)</span>
<span id="cb98-23"><a href="#cb98-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-24"><a href="#cb98-24" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/opt/conda/envs/py38/lib/python3.8/site-packages/statsmodels/robust/robust_linear_model.py:288: ConvergenceWarning: Estimated scale is 0.0 indicating that the most last iteration produced a perfect fit of the weighted data.
  warnings.warn('Estimated scale is 0.0 indicating that the most'</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="prots2prot-thiolase-codehidden_files/figure-html/cell-64-output-2.png" class="img-fluid"></p>
</div>
</div>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>